// @useparagon/core/dist/inputs/input.constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INPUT_NAME_TO_INPUT_TYPE_MAP = exports.BASIC_INPUT_NAME_TO_INPUT_TYPE_MAP = void 0;
const input_interface_1 = require("./input.interface");
exports.BASIC_INPUT_NAME_TO_INPUT_TYPE_MAP = {
    text: input_interface_1.SidebarInputType.ValueText,
    number: input_interface_1.SidebarInputType.Number,
    email: input_interface_1.SidebarInputType.Email,
    url: input_interface_1.SidebarInputType.URL,
    boolean: input_interface_1.SidebarInputType.BooleanInput,
    password: input_interface_1.SidebarInputType.Password,
    custom_dropdown: input_interface_1.SidebarInputType.CustomDropdown,
};
exports.INPUT_NAME_TO_INPUT_TYPE_MAP = {
    ...exports.BASIC_INPUT_NAME_TO_INPUT_TYPE_MAP,
    [input_interface_1.DataSourceType.DYNAMIC]: input_interface_1.SidebarInputType.DynamicEnum,
    [input_interface_1.DataSourceType.FIELD_MAPPER]: input_interface_1.SidebarInputType.FieldMapper,
    [input_interface_1.DataSourceType.COMBO_INPUT]: input_interface_1.SidebarInputType.ComboInput,
    [input_interface_1.DataSourceType.STATIC_ENUM]: input_interface_1.SidebarInputType.Enum,
};
