// @useparagon/core/dist/inputs/input.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInputs = exports.SidebarInputType = exports.DataSourceType = void 0;
var DataSourceType;
(function (DataSourceType) {
    DataSourceType["DYNAMIC"] = "DYNAMIC_DATA_SOURCE";
    DataSourceType["STATIC_ENUM"] = "STATIC_ENUM_DATA_SOURCE";
    DataSourceType["FIELD_MAPPER"] = "FIELD_MAPPER_DATA_SOURCE";
    DataSourceType["COMBO_INPUT"] = "COMBO_INPUT_DATA_SOURCE";
    DataSourceType["DYNAMIC_COMBO_INPUT"] = "DYNAMIC_COMBO_INPUT_DATA_SOURCE";
})(DataSourceType || (exports.DataSourceType = DataSourceType = {}));
var SidebarInputType;
(function (SidebarInputType) {
    SidebarInputType["BooleanInput"] = "BOOLEAN_INPUT";
    SidebarInputType["Code"] = "CODE";
    SidebarInputType["ComboInput"] = "COMBO_INPUT";
    SidebarInputType["Conditional"] = "CONDITIONAL";
    SidebarInputType["CustomDropdown"] = "CUSTOM_DROPDOWN";
    SidebarInputType["CustomFieldMapping"] = "CUSTOM_FIELD_MAPPING";
    SidebarInputType["DynamicComboInput"] = "DYNAMIC_COMBO_INPUT";
    SidebarInputType["DynamicConditional"] = "DYNAMIC_CONDITIONAL";
    SidebarInputType["DynamicEnum"] = "DYNAMIC_ENUM";
    SidebarInputType["EditableEnum"] = "EDITABLE_ENUM";
    SidebarInputType["Email"] = "EMAIL";
    SidebarInputType["Enum"] = "ENUM";
    SidebarInputType["FieldMapper"] = "FIELD_MAPPER";
    SidebarInputType["Number"] = "NUMBER";
    SidebarInputType["Password"] = "PASSWORD";
    SidebarInputType["Switch"] = "SWITCH";
    SidebarInputType["Text"] = "TEXT";
    SidebarInputType["TextArea"] = "TEXTAREA";
    SidebarInputType["URL"] = "URL";
    SidebarInputType["ValueText"] = "TEXT_NO_VARS";
    SidebarInputType["ValueTextArea"] = "TEXTAREA_NO_VARS";
})(SidebarInputType || (exports.SidebarInputType = SidebarInputType = {}));
const createInputs = (inputs) => inputs;
exports.createInputs = createInputs;
