// @useparagon/core/dist/steps/library/function/function.step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionStep = void 0;
const execution_1 = require("../../../execution");
const resolver_utils_1 = require("../../../resolvers/resolver.utils");
const step_1 = require("../../step");
const step_interface_1 = require("../../step.interface");
class FunctionStep extends step_1.Step {
    type = step_interface_1.StepType.FUNCTION;
    parameters;
    get output() {
        return (0, execution_1.createProxyInputReference)(this, () => this.id, execution_1.ProxyVariableSourceType.STEP_REFERENCE);
    }
    _parameters;
    constructor(params) {
        super(params);
        this._parameters = params;
        this.parameters = this.serializeParameters();
    }
    serializeParameters() {
        const initParams = this._parameters;
        return {
            parameters: (0, resolver_utils_1.resolveParamsToSources)(initParams.parameters, undefined, false, this.dependencies),
            code: initParams.code.toString(),
            retryOnFailure: Boolean(initParams.autoRetry),
        };
    }
}
exports.FunctionStep = FunctionStep;
