// @useparagon/integrations/linkedin/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for linkedin
 */
exports.inputs = [];

/**
 * triggers for linkedin
 */
exports.triggers = [];

/**
 * intents for linkedin
 */
exports.intents = [
  {
    title: 'Get Profile By ID',
    name: 'LINKEDIN_GET_PROFILE_BY_ID',
    output: [],
    inputs: [
      {
        id: 'id',
        title: 'Profile ID',
        type: 'TEXTAREA',
        subtitle: 'Specify the ID of the profile to fetch.',
        required: true,
      },
    ],
  },
  {
    title: 'Create Post',
    name: 'LINKEDIN_CREATE_POST',
    output: [],
    inputs: [
      {
        id: 'author',
        title: 'Author',
        type: 'TEXTAREA',
        subtitle: 'Specify the Author who is creating the post.',
        placeholder: 'person:-OLWJgnReo',
        required: true,
      },
      {
        id: 'lifecycleState',
        title: 'Life Cycle State',
        type: 'ENUM',
        subtitle:
          'Specify the state of the post, you can use "PUBLISHED" to make it visible to others.',
        required: true,
        values: ['PUBLISHED', 'DRAFT'],
      },
      {
        id: 'text',
        title: 'Text',
        type: 'TEXTAREA',
        subtitle: 'Specify the text of the post.',
        required: true,
      },
      {
        id: 'visibility',
        title: 'Visibility',
        type: 'ENUM',
        subtitle: 'Specify the visibility of the post.',
        required: true,
        values: ['CONNECTIONS', 'PUBLIC'],
      },
    ],
  },
];

/**
 * integration class for linkedin
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'linkedin',

  /**
   * action step methods
   */
  actions: {
    /**
     * Get Profile By ID
     * method for LINKEDIN_GET_PROFILE_BY_ID
     */
    getProfileById: (inputs, options) => {
      return new ActionStep({
        intent: 'LINKEDIN_GET_PROFILE_BY_ID',
        actionType: 'linkedin',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Create Post
     * method for LINKEDIN_CREATE_POST
     */
    createPost: (inputs, options) => {
      return new ActionStep({
        intent: 'LINKEDIN_CREATE_POST',
        actionType: 'linkedin',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {},
};
