// @useparagon/core/dist/steps/step.interface-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRIGGER_TYPES = exports.StepType = void 0;
var StepType;
(function (StepType) {
    StepType["ACTION"] = "ACTION/CUSTOM";
    StepType["ACTION_TRIGGER"] = "TRIGGER/ACTION";
    StepType["CRON"] = "TRIGGER/CRON";
    StepType["CUSTOM_INTEGRATION_REQUEST"] = "ACTION/CUSTOM_INTEGRATION_REQUEST";
    StepType["DELAY"] = "ACTION/DELAY";
    StepType["ENDPOINT"] = "TRIGGER/ENDPOINT";
    StepType["EVENT"] = "TRIGGER/EVENT";
    StepType["FUNCTION"] = "ACTION/FUNCTION";
    StepType["IFELSE"] = "TRANSITION/IFELSE";
    StepType["INTEGRATION_ENABLED"] = "TRIGGER/INTEGRATION_ENABLED";
    StepType["MAP"] = "TRANSITION/MAP";
    StepType["OAUTH"] = "TRIGGER/OAUTH";
    StepType["REDIRECT"] = "ACTION/REDIRECT";
    StepType["RESPONSE"] = "ACTION/RESPONSE";
    StepType["REQUEST"] = "ACTION/REQUEST";
    StepType["UNSELECTED_TRIGGER"] = "TRIGGER/NONE";
})(StepType || (exports.StepType = StepType = {}));
exports.TRIGGER_TYPES = [
    StepType.UNSELECTED_TRIGGER,
    StepType.CRON,
    StepType.ENDPOINT,
    StepType.OAUTH,
    StepType.ACTION_TRIGGER,
    StepType.EVENT,
    StepType.INTEGRATION_ENABLED,
];
