// @useparagon/integrations/pipedrive/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for pipedrive
 */
exports.inputs = [
  {
    id: 'cachePipelines',
    title: 'Pipeline (Deprecated)',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachePipelinesV2',
    title: 'Pipeline',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cacheStages',
    title: 'Stage (Deprecated)',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cacheStagesV2',
    title: 'Stage',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'customObjectMapping',
    title: 'Custom Field Mapping',
    type: 'FIELD_MAPPER_DATA_SOURCE',
    fieldSource: {
      id: 'cacheObjectFields',
      title: 'Custom Object Field',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    recordSource: {
      id: 'cacheObjectTypes',
      title: 'Custom Object Types',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
];

/**
 * triggers for pipedrive
 */
exports.triggers = [
  {
    title: 'Record Created',
    name: 'PIPEDRIVE_TRIGGER_RECORD_CREATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          'activities',
          'activities_v2',
          'deals',
          'deals_v2',
          'organizations',
          'organizations_v2',
          'persons',
          'persons_v2',
          'notes',
        ],
      },
      {
        id: 'triggerFilterFormula',
        title: 'Filter records',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on records that match these filters.',
        supportedOperators: [
          '$arrayIsEmpty',
          '$arrayIsNotEmpty',
          '$booleanFalse',
          '$booleanTrue',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$isUndefinedOrNull',
          '$isNotUndefinedOrNull',
          '$exists',
          '$doesNotExist',
          '$numberDoesNotEqual',
          '$numberEquals',
          '$numberGreaterThan',
          '$numberLessThan',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringDoesNotEndWith',
          '$stringDoesNotExactlyMatch',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringExactlyMatches',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
        ],
      },
    ],
  },
  {
    title: 'Record Updated',
    name: 'PIPEDRIVE_TRIGGER_RECORD_UPDATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          'activities',
          'activities_v2',
          'deals',
          'deals_v2',
          'organizations',
          'organizations_v2',
          'persons',
          'persons_v2',
          'notes',
        ],
      },
      {
        id: 'triggerFilterFormula',
        title: 'Filter records',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on records that match these filters.',
        supportedOperators: [
          '$arrayIsEmpty',
          '$arrayIsNotEmpty',
          '$booleanFalse',
          '$booleanTrue',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$isUndefinedOrNull',
          '$isNotUndefinedOrNull',
          '$exists',
          '$doesNotExist',
          '$numberDoesNotEqual',
          '$numberEquals',
          '$numberGreaterThan',
          '$numberLessThan',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringDoesNotEndWith',
          '$stringDoesNotExactlyMatch',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringExactlyMatches',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
        ],
      },
    ],
  },
  {
    title: 'Lead Created',
    name: 'PIPEDRIVE_TRIGGER_LEAD_CREATED',
    output: [],
    inputs: [],
  },
];

/**
 * intents for pipedrive
 */
exports.intents = [
  {
    title: 'Create Record',
    name: 'PIPEDRIVE_CREATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'activities',
            dependentInputs: [
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
              },
              {
                id: 'note',
                title: 'Note',
                type: 'TEXTAREA',
              },
              {
                id: 'type',
                title: 'Type',
                type: 'ENUM',
                subtitle: 'Defaults to Call if left blank.',
                values: [
                  'call',
                  'meeting',
                  'task',
                  'deadline',
                  'email',
                  'lunch',
                ],
              },
              {
                id: 'dueDate',
                title: 'Due Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'location',
                title: 'Location',
                type: 'TEXTAREA',
                subtitle: 'The address of the activity.',
              },
              {
                id: 'done',
                title: 'Done',
                type: 'BOOLEAN_INPUT',
              },
              {
                id: 'userId',
                title: 'User ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user whom the activity is assigned to. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'dealId',
                title: 'Deal ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the deal this activity is associated with.',
              },
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Lead this activity is associated with.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Person this activity is associated with.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Organization this activity is associated with.',
              },
            ],
          },
          {
            value: 'activities_v2',
            dependentInputs: [
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
              },
              {
                id: 'note',
                title: 'Note',
                type: 'TEXTAREA',
              },
              {
                id: 'type',
                title: 'Type',
                type: 'ENUM',
                subtitle: 'Defaults to Call if left blank.',
                values: [
                  'call',
                  'meeting',
                  'task',
                  'deadline',
                  'email',
                  'lunch',
                ],
              },
              {
                id: 'dueDate',
                title: 'Due Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'location',
                title: 'Location',
                type: 'TEXTAREA',
                subtitle: 'The address of the activity.',
              },
              {
                id: 'done',
                title: 'Done',
                type: 'BOOLEAN_INPUT',
              },
              {
                id: 'userId',
                title: 'User ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user whom the activity is assigned to. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'dealId',
                title: 'Deal ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the deal this activity is associated with.',
              },
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Lead this activity is associated with.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who the deal will be the owner of the lead. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Person this activity is associated with.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Organization this activity is associated with.',
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'value',
                title: 'Value',
                type: 'TEXTAREA',
              },
              {
                id: 'status',
                title: 'Status',
                type: 'ENUM',
                subtitle: 'Defaults to Open if left blank.',
                defaultValue: 'open',
                values: ['open', 'won', 'lost', 'deleted'],
              },
              {
                id: 'expectedCloseDate',
                title: 'Expected Close Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'userId',
                title: 'User ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who the deal will be assigned to. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the person who the deal will be assigned to.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of an organization which this deal will be linked to.',
              },
              {
                id: 'pipelineId',
                title: 'Pipeline ID',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select which pipeline the deal will be added to. By default, the deal will be added to the first stage of the specified pipeline.',
              },
              {
                id: 'stageId',
                title: 'Stage ID',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select which stage the deal will be added to. A pipeline will be assigned automatically based on the Stage ID. If left blank, the deal will be placed in the first stage of the default pipeline.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which deal fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'deals_v2',
            dependentInputs: [
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'value',
                title: 'Value',
                type: 'TEXTAREA',
              },
              {
                id: 'status',
                title: 'Status',
                type: 'ENUM',
                subtitle: 'Defaults to Open if left blank.',
                defaultValue: 'open',
                values: ['open', 'won', 'lost', 'deleted'],
              },
              {
                id: 'expectedCloseDate',
                title: 'Expected Close Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who the deal will be assigned to. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the person who the deal will be assigned to.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of an organization which this deal will be linked to.',
              },
              {
                id: 'pipelineId',
                title: 'Pipeline ID',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select which pipeline the deal will be added to. By default, the deal will be added to the first stage of the specified pipeline.',
              },
              {
                id: 'stageId',
                title: 'Stage ID',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select which stage the deal will be added to. A pipeline will be assigned automatically based on the Stage ID. If left blank, the deal will be placed in the first stage of the default pipeline.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which deal fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'organizations',
            dependentInputs: [
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who will be marked as the owner of the organization. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which organization fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'organizations_v2',
            dependentInputs: [
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who will be marked as the owner of the organization. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which organization fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'persons',
            dependentInputs: [
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the organization this person will belong to.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who will be marked as the owner of the person. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which persons fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'persons_v2',
            dependentInputs: [
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'emails',
                title: 'Emails',
                type: 'TEXTAREA',
                placeholder:
                  '[{"value": "email@example.com", "primary": true, "label": "Work"}]',
              },
              {
                id: 'phones',
                title: 'Phones',
                type: 'TEXTAREA',
                placeholder:
                  '[{"value": "+1234567890", "primary": true, "label": "Work"}]',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the organization this person will belong to.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who will be marked as the owner of the person. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which persons fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'leads',
            dependentInputs: [
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'value',
                title: 'Value',
                type: 'TEXTAREA',
              },
              {
                id: 'expectedCloseDate',
                title: 'Expected Close Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who the deal will be the owner of the lead. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of a person which this lead will be linked to. This property is required unless Organization ID is specified.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of an organization which this lead will be linked to. This property is required unless Person ID is specified.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which lead fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'notes',
            dependentInputs: [
              {
                id: 'content',
                title: 'Content',
                type: 'TEXTAREA',
                subtitle: 'The content of the note. Supports HTML.',
                required: true,
              },
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the lead the note will be attached to. Required unless Person ID, Organization ID, or Deal ID is provided.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the person the note will be attached to. Required unless Lead ID, Organization ID, or Deal ID is provided.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the organization the note will be attached to. Required unless Lead ID, Person ID, or Deal ID is provided.',
              },
              {
                id: 'dealId',
                title: 'Deal ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Deal the note will be attached to. Required unless Lead ID, Person ID, or Organization ID is provided.',
              },
            ],
          },
          {
            value: 'users',
            dependentInputs: [
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'email',
                title: 'Email',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'activeStatus',
                title: 'Active Status',
                type: 'ENUM',
                required: true,
                defaultValue: 'true',
                values: ['true', 'false'],
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Update Record',
    name: 'PIPEDRIVE_UPDATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'activities',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
              },
              {
                id: 'note',
                title: 'Note',
                type: 'TEXTAREA',
              },
              {
                id: 'type',
                title: 'Type',
                type: 'ENUM',
                values: [
                  'call',
                  'meeting',
                  'task',
                  'deadline',
                  'email',
                  'lunch',
                ],
              },
              {
                id: 'dueDate',
                title: 'Due Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'location',
                title: 'Location',
                type: 'TEXTAREA',
                subtitle: 'The address of the activity.',
              },
              {
                id: 'done',
                title: 'Done',
                type: 'BOOLEAN_INPUT',
              },
              {
                id: 'userId',
                title: 'User ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user whom the activity is assigned to.',
              },
              {
                id: 'dealId',
                title: 'Deal ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the deal this activity is associated with.',
              },
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Lead this activity is associated with.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Person this activity is associated with.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Organization this activity is associated with.',
              },
            ],
          },
          {
            value: 'activities_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
              },
              {
                id: 'note',
                title: 'Note',
                type: 'TEXTAREA',
              },
              {
                id: 'type',
                title: 'Type',
                type: 'ENUM',
                values: [
                  'call',
                  'meeting',
                  'task',
                  'deadline',
                  'email',
                  'lunch',
                ],
              },
              {
                id: 'dueDate',
                title: 'Due Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'location',
                title: 'Location',
                type: 'TEXTAREA',
                subtitle: 'The address of the activity.',
              },
              {
                id: 'done',
                title: 'Done',
                type: 'BOOLEAN_INPUT',
              },
              {
                id: 'userId',
                title: 'User ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user whom the activity is assigned to.',
              },
              {
                id: 'dealId',
                title: 'Deal ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the deal this activity is associated with.',
              },
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Lead this activity is associated with.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who the deal will be the owner of the lead. Defaults to the currently authenticated user if left blank.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Person this activity is associated with.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Organization this activity is associated with.',
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the deal that will be updated.',
                required: true,
              },
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
              },
              {
                id: 'value',
                title: 'Value',
                type: 'TEXTAREA',
              },
              {
                id: 'status',
                title: 'Status',
                type: 'ENUM',
                subtitle: 'Defaults to Open if left blank.',
                defaultValue: 'open',
                values: ['open', 'won', 'lost', 'deleted'],
              },
              {
                id: 'expectedCloseDate',
                title: 'Expected Close Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'userId',
                title: 'User ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who the deal will be assigned to.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the person who the deal will be assigned to.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of an organization which this deal will be linked to.',
              },
              {
                id: 'pipelineId',
                title: 'Pipeline ID',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select which pipeline the deal will be added to.',
              },
              {
                id: 'stageId',
                title: 'Stage ID',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select which stage the deal will be added to. A pipeline will be assigned automatically based on the Stage ID.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which deal fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'deals_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the deal that will be updated.',
                required: true,
              },
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
              },
              {
                id: 'value',
                title: 'Value',
                type: 'TEXTAREA',
              },
              {
                id: 'status',
                title: 'Status',
                type: 'ENUM',
                subtitle: 'Defaults to Open if left blank.',
                defaultValue: 'open',
                values: ['open', 'won', 'lost', 'deleted'],
              },
              {
                id: 'expectedCloseDate',
                title: 'Expected Close Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who the deal will be assigned to.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the person who the deal will be assigned to.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of an organization which this deal will be linked to.',
              },
              {
                id: 'pipelineId',
                title: 'Pipeline ID',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select which pipeline the deal will be added to.',
              },
              {
                id: 'stageId',
                title: 'Stage ID',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select which stage the deal will be added to. A pipeline will be assigned automatically based on the Stage ID.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which deal fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'organizations',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the organization that will be updated.',
                required: true,
              },
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who will be marked as the owner of the organization.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which organization fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'organizations_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the organization that will be updated.',
                required: true,
              },
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who will be marked as the owner of the organization.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which organization fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'persons',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the person that will be updated.',
                required: true,
              },
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
              },
              {
                id: 'email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the organization this person will belong to.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who will be marked as the owner of the person.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which persons fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'persons_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the person that will be updated.',
                required: true,
              },
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
              },
              {
                id: 'emails',
                title: 'Emails',
                type: 'TEXTAREA',
                placeholder:
                  '[{"value": "email@example.com", "primary": true, "label": "Work"}]',
              },
              {
                id: 'phones',
                title: 'Phones',
                type: 'TEXTAREA',
                placeholder:
                  '[{"value": "+1234567890", "primary": true, "label": "Work"}]',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the organization this person will belong to.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who will be marked as the owner of the person.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which persons fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'leads',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the lead that will be updated.',
                required: true,
              },
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
              },
              {
                id: 'value',
                title: 'Value',
                type: 'TEXTAREA',
              },
              {
                id: 'expectedCloseDate',
                title: 'Expected Close Date',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO8601 date formats.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the user who the deal will be the owner of the lead.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of a person which this lead will be linked to.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of an organization which this lead will be linked to.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which lead fields to update.',
                placeholder:
                  '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'notes',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the note that will be updated.',
                required: true,
              },
              {
                id: 'content',
                title: 'Content',
                type: 'TEXTAREA',
                subtitle: 'The content of the note. Supports HTML.',
                required: true,
              },
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the lead the note will be attached to.',
              },
              {
                id: 'personId',
                title: 'Person ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the person the note will be attached to.',
              },
              {
                id: 'organizationId',
                title: 'Organization ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the organization the note will be attached to.',
              },
              {
                id: 'dealId',
                title: 'Deal ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Deal the note will be attached to.',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get Record by ID',
    name: 'PIPEDRIVE_GET_RECORD_BY_ID',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'activities',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'activities_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'deals_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'organizations',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'organizations_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'persons',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'persons_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'leads',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'notes',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
          {
            value: 'users',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the activity that will be updated.',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get Records',
    name: 'PIPEDRIVE_GET_RECORDS',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'term',
                title: 'Term',
                type: 'TEXTAREA',
                subtitle: 'The search term to look for.',
                required: true,
              },
              {
                id: 'recordFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for items that match specified filters.',
                supportedKeys: ['person_id', 'organization_id'],
                supportedOperators: ['$numberEquals', '$stringExactlyMatches'],
              },
            ],
          },
          {
            value: 'deals_v2',
            dependentInputs: [
              {
                id: 'term',
                title: 'Term',
                type: 'TEXTAREA',
                subtitle: 'The search term to look for.',
                required: true,
              },
              {
                id: 'recordFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for items that match specified filters.',
                supportedKeys: ['person_id', 'organization_id'],
                supportedOperators: ['$numberEquals', '$stringExactlyMatches'],
              },
            ],
          },
          {
            value: 'organizations',
            dependentInputs: [
              {
                id: 'term',
                title: 'Term',
                type: 'TEXTAREA',
                subtitle: 'The search term to look for.',
                required: true,
              },
            ],
          },
          {
            value: 'organizations_v2',
            dependentInputs: [
              {
                id: 'term',
                title: 'Term',
                type: 'TEXTAREA',
                subtitle: 'The search term to look for.',
                required: true,
              },
              {
                id: 'recordFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for items that match specified filters.',
                supportedKeys: ['person_id', 'organization_id'],
                supportedOperators: ['$numberEquals', '$stringExactlyMatches'],
              },
            ],
          },
          {
            value: 'persons',
            dependentInputs: [
              {
                id: 'term',
                title: 'Term',
                type: 'TEXTAREA',
                subtitle: 'The search term to look for.',
                required: true,
              },
              {
                id: 'recordFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for items that match specified filters.',
                supportedKeys: ['person_id', 'organization_id'],
                supportedOperators: ['$numberEquals', '$stringExactlyMatches'],
              },
            ],
          },
          {
            value: 'persons_v2',
            dependentInputs: [
              {
                id: 'term',
                title: 'Term',
                type: 'TEXTAREA',
                subtitle: 'The search term to look for.',
                required: true,
              },
              {
                id: 'recordFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for items that match specified filters.',
                supportedKeys: ['person_id', 'organization_id'],
                supportedOperators: ['$numberEquals', '$stringExactlyMatches'],
              },
            ],
          },
          {
            value: 'leads',
            dependentInputs: [
              {
                id: 'term',
                title: 'Term',
                type: 'TEXTAREA',
                subtitle: 'The search term to look for.',
                required: true,
              },
              {
                id: 'recordFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for items that match specified filters.',
                supportedKeys: ['person_id', 'organization_id'],
                supportedOperators: ['$numberEquals', '$stringExactlyMatches'],
              },
            ],
          },
          {
            value: 'leads_v2',
            dependentInputs: [
              {
                id: 'term',
                title: 'Term',
                type: 'TEXTAREA',
                subtitle: 'The search term to look for.',
                required: true,
              },
              {
                id: 'recordFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for items that match specified filters.',
                supportedKeys: ['person_id', 'organization_id'],
                supportedOperators: ['$numberEquals', '$stringExactlyMatches'],
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Delete Record',
    name: 'PIPEDRIVE_DELETE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'activities',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'activities_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'deals_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'organizations',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'organizations_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'persons',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'persons_v2',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'leads',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'notes',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'users',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the event to delete.',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
];

/**
 * integration class for pipedrive
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'pipedrive',

  /**
   * action step methods
   */
  actions: {
    /**
     * Create Record
     * method for PIPEDRIVE_CREATE_RECORD
     */
    createRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'PIPEDRIVE_CREATE_RECORD',
        actionType: 'pipedrive',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update Record
     * method for PIPEDRIVE_UPDATE_RECORD
     */
    updateRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'PIPEDRIVE_UPDATE_RECORD',
        actionType: 'pipedrive',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Record by ID
     * method for PIPEDRIVE_GET_RECORD_BY_ID
     */
    getRecordById: (inputs, options) => {
      return new ActionStep({
        intent: 'PIPEDRIVE_GET_RECORD_BY_ID',
        actionType: 'pipedrive',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Records
     * method for PIPEDRIVE_GET_RECORDS
     */
    getRecords: (inputs, options) => {
      return new ActionStep({
        intent: 'PIPEDRIVE_GET_RECORDS',
        actionType: 'pipedrive',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Delete Record
     * method for PIPEDRIVE_DELETE_RECORD
     */
    deleteRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'PIPEDRIVE_DELETE_RECORD',
        actionType: 'pipedrive',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {
    /**
     * Record Created
     * method for PIPEDRIVE_TRIGGER_RECORD_CREATED
     */
    recordCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'PIPEDRIVE_TRIGGER_RECORD_CREATED',
        actionType: 'pipedrive',
        parameters: inputs,
      });
    },
    /**
     * Record Updated
     * method for PIPEDRIVE_TRIGGER_RECORD_UPDATED
     */
    recordUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'PIPEDRIVE_TRIGGER_RECORD_UPDATED',
        actionType: 'pipedrive',
        parameters: inputs,
      });
    },
    /**
     * Lead Created
     * method for PIPEDRIVE_TRIGGER_LEAD_CREATED
     */
    leadCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'PIPEDRIVE_TRIGGER_LEAD_CREATED',
        actionType: 'pipedrive',
        parameters: inputs,
      });
    },
  },
};
