// @useparagon/integrations/marketo/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for marketo
 */
exports.inputs = [
  {
    id: 'lists',
    title: 'List',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'customObjectMapping',
    title: 'Field Mapping',
    type: 'FIELD_MAPPER_DATA_SOURCE',
    fieldSource: {
      id: 'customObjectFields',
      title: 'Custom Object fields',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    recordSource: {
      id: 'customObjects',
      title: 'Custom Objects',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
];

/**
 * triggers for marketo
 */
exports.triggers = [
  {
    title: 'Lead Added to a list',
    name: 'MARKETO_TRIGGER_LEAD_ADDED_TO_LIST',
    output: [],
    inputs: [
      {
        id: 'listId',
        title: 'List ID',
        type: 'TEXTAREA',
        subtitle: 'Only trigger for leads associated with this list.',
        required: true,
      },
    ],
  },
];

/**
 * intents for marketo
 */
exports.intents = [
  {
    title: 'Create or Update Lead',
    name: 'MARKETO_CREATE_OR_UPDATE_LEAD',
    output: [],
    inputs: [
      {
        id: 'email',
        title: 'Email',
        type: 'TEXTAREA',
        subtitle:
          'If there already exists a lead with this email, it will be updated.',
        required: true,
      },
      {
        id: 'firstName',
        title: 'First Name',
        type: 'TEXTAREA',
      },
      {
        id: 'lastName',
        title: 'Last Name',
        type: 'TEXTAREA',
      },
      {
        id: 'company',
        title: 'Company',
        type: 'TEXTAREA',
      },
      {
        id: 'title',
        title: 'Title',
        type: 'TEXTAREA',
      },
      {
        id: 'phone',
        title: 'Phone',
        type: 'TEXTAREA',
      },
      {
        id: 'website',
        title: 'Website',
        type: 'TEXTAREA',
      },
      {
        id: 'additionalFields',
        title: 'Additional Fields',
        type: 'CODE',
        subtitle:
          'Specify any other fields that should be updated in JSON below.',
        placeholder:
          '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
      },
    ],
  },
  {
    title: 'Get Leads',
    name: 'MARKETO_GET_LEADS',
    output: [],
    inputs: [
      {
        id: 'filterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Search for records that match specified filters.',
        supportedKeys: [
          'cookies',
          'department',
          'email',
          'mktoName',
          'id',
          'leadPartitionId',
        ],
        supportedOperators: ['$stringIsIn'],
      },
    ],
  },
  {
    title: 'Get Lead by ID',
    name: 'MARKETO_GET_LEAD_BY_ID',
    output: [],
    inputs: [
      {
        id: 'leadId',
        title: 'Lead ID',
        type: 'TEXTAREA',
        required: true,
      },
    ],
  },
  {
    title: 'Add Leads to List',
    name: 'MARKETO_ADD_LEADS_TO_LIST',
    output: [],
    inputs: [
      {
        id: 'leadIds',
        title: 'Lead IDs',
        type: 'TEXTAREA',
        subtitle:
          'IDs of leads to be added to the List (up to 300). You can enter a single ID or an array of IDs.',
        required: true,
      },
      {
        id: 'listId',
        title: 'List',
        type: 'TEXTAREA',
        subtitle:
          'The List that the lead(s) should be added to. Use Connect Portal Workflow Settings to allow users to select a List. ',
        placeholder: '{{settings.list}}',
      },
    ],
  },
  {
    title: 'Create Custom Object',
    name: 'MARKETO_CREATE_CUSTOM_OBJECT',
    output: [],
    inputs: [
      {
        id: 'customObjectName',
        title: 'Object Name',
        type: 'TEXTAREA',
        subtitle:
          'The API Name of the Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
        placeholder: '{{settings.objectType}}',
        required: true,
      },
      {
        id: 'recordFields',
        title: 'Record Fields',
        type: 'CODE',
        subtitle:
          'Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping to specify which fields to update.',
        placeholder:
          '{\n"property_number": "17",\n"property_dropdown": "choice_b",\n"property_radio": "option_1",\n"property_string": "value",\n"property_date": "1572480000000"\n}',
        required: true,
      },
    ],
  },
];

/**
 * integration class for marketo
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'marketo',

  /**
   * action step methods
   */
  actions: {
    /**
     * Create or Update Lead
     * method for MARKETO_CREATE_OR_UPDATE_LEAD
     */
    createOrUpdateLead: (inputs, options) => {
      return new ActionStep({
        intent: 'MARKETO_CREATE_OR_UPDATE_LEAD',
        actionType: 'marketo',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Leads
     * method for MARKETO_GET_LEADS
     */
    getLeads: (inputs, options) => {
      return new ActionStep({
        intent: 'MARKETO_GET_LEADS',
        actionType: 'marketo',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Lead by ID
     * method for MARKETO_GET_LEAD_BY_ID
     */
    getLeadById: (inputs, options) => {
      return new ActionStep({
        intent: 'MARKETO_GET_LEAD_BY_ID',
        actionType: 'marketo',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Add Leads to List
     * method for MARKETO_ADD_LEADS_TO_LIST
     */
    addLeadsToList: (inputs, options) => {
      return new ActionStep({
        intent: 'MARKETO_ADD_LEADS_TO_LIST',
        actionType: 'marketo',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Create Custom Object
     * method for MARKETO_CREATE_CUSTOM_OBJECT
     */
    createCustomObject: (inputs, options) => {
      return new ActionStep({
        intent: 'MARKETO_CREATE_CUSTOM_OBJECT',
        actionType: 'marketo',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {
    /**
     * Lead Added to a list
     * method for MARKETO_TRIGGER_LEAD_ADDED_TO_LIST
     */
    leadAddedToList: (inputs) => {
      return new ActionTriggerStep({
        intent: 'MARKETO_TRIGGER_LEAD_ADDED_TO_LIST',
        actionType: 'marketo',
        parameters: inputs,
      });
    },
  },
};
