// @useparagon/integrations/microsoftDynamics/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for microsoftDynamics
 */
exports.inputs = [
  {
    id: 'cachedCustomEntities',
    title: 'Custom Entity Name (Legacy)',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedCustomEntitySetName',
    title: 'Custom Entity Set Name',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedCustomers',
    title: 'Potential customer ID',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedOwners',
    title: 'Owner (User or Team)',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedAccounts',
    title: 'Account',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedContacts',
    title: 'Contact',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedLeads',
    title: 'Lead',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedOpportunities',
    title: 'Opportunity',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedProcessStages',
    title: 'Process Stage',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'cachedTransactionCurrencies',
    title: 'Transaction Currency',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'customObjectMapping',
    title: 'Field Mapping',
    type: 'FIELD_MAPPER_DATA_SOURCE',
    fieldSource: {
      id: 'cachedCustomObjectFields',
      title: 'Custom object Field',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    recordSource: {
      id: 'cachedCustomEntities',
      title: 'Custom Entity Name (Legacy)',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
];

/**
 * triggers for microsoftDynamics
 */
exports.triggers = [
  {
    title: 'New Record',
    name: 'MICROSOFT_DYNAMICS_TRIGGER_RECORD_CREATED',
    output: [],
    inputs: [
      {
        id: 'mdEntity',
        title: 'Choose an entity',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'name',
                  'customerid',
                  'salesstagecode',
                  'estimatedclosedate',
                  'actualclosedate',
                  'estimatedvalue',
                  'description',
                  '_ownerid_value',
                  'transactioncurrencyid',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'name',
                  '_ownerid_value',
                  'websiteurl',
                  'telephone1',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'firstname',
                  'lastname',
                  'emailaddress1',
                  'telephone1',
                  'jobtitle',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'firstname',
                  'lastname',
                  'emailaddress1',
                  'telephone1',
                  'websiteurl',
                  'jobtitle',
                  'salesstagecode',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Call',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'directioncode',
                  '_regardingobjectid_value',
                  'phonenumber',
                  'actualdurationminutes',
                  'description',
                  'scheduledend',
                  'prioritycode',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Meeting',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'scheduledstart',
                  'scheduledend',
                  '_regardingobjectid_value',
                  'location',
                  'description',
                  'prioritycode',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  '_regardingobjectid_value',
                  'scheduledend',
                  '_ownerid_value',
                  'prioritycode',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Note',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: ['_objectid_value', 'subject', 'notetext'],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: '<specified by entity property>',
            dependentInputs: [
              {
                id: 'customMdEntityName',
                title: 'Entity Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Entity Set Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: 'entitysetname',
                required: true,
              },
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Record Updated',
    name: 'MICROSOFT_DYNAMICS_TRIGGER_RECORD_UPDATED',
    output: [],
    inputs: [
      {
        id: 'mdEntity',
        title: 'Choose an entity',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'name',
                  'customerid',
                  'salesstagecode',
                  'estimatedclosedate',
                  'actualclosedate',
                  'estimatedvalue',
                  'description',
                  '_ownerid_value',
                  'transactioncurrencyid',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'name',
                  '_ownerid_value',
                  'websiteurl',
                  'telephone1',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'firstname',
                  'lastname',
                  'emailaddress1',
                  'telephone1',
                  'jobtitle',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'firstname',
                  'lastname',
                  'emailaddress1',
                  'telephone1',
                  'websiteurl',
                  'jobtitle',
                  'salesstagecode',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Call',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'directioncode',
                  '_regardingobjectid_value',
                  'phonenumber',
                  'actualdurationminutes',
                  'description',
                  'scheduledend',
                  'prioritycode',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Meeting',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'scheduledstart',
                  'scheduledend',
                  '_regardingobjectid_value',
                  'location',
                  'description',
                  'prioritycode',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  '_regardingobjectid_value',
                  'scheduledend',
                  '_ownerid_value',
                  'prioritycode',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Note',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: ['_objectid_value', 'subject', 'notetext'],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: '<specified by entity property>',
            dependentInputs: [
              {
                id: 'customMdEntityName',
                title: 'Entity Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Entity Set Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: 'entitysetname',
                required: true,
              },
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
        ],
      },
    ],
  },
];

/**
 * intents for microsoftDynamics
 */
exports.intents = [
  {
    title: 'Create record',
    name: 'MICROSOFT_DYNAMICS_CREATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'mdEntity',
        title: 'Choose an entity',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                subtitle:
                  'A subject or descriptive name, such as the expected order or company name, for the Opportunity. This field is required.',
                required: true,
              },
              {
                id: 'customerid',
                title: 'Potential Customer ID',
                type: 'TEXTAREA',
                subtitle:
                  'Select an account or contact to associate with this Opportunity. This field is required.',
                placeholder: '{{settings.customerID}}',
                required: true,
              },
              {
                id: 'salesstagecode',
                title: 'Process Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Process Stage.',
                placeholder: '{{settings.processStage}}',
              },
              {
                id: 'estimatedclosedate',
                title: 'Estimated Close Date',
                type: 'TEXTAREA',
                subtitle: 'The expected closing date of the Opportunity.',
              },
              {
                id: 'actualclosedate',
                title: 'Actual Close Date',
                type: 'TEXTAREA',
                subtitle:
                  'The date and time when the Opportunity was closed or canceled.',
              },
              {
                id: 'estimatedvalue',
                title: 'Est. Revenue',
                type: 'TEXTAREA',
                subtitle:
                  'The estimated revenue amount to indicate the potential sale or value of the opportunity for revenue forecasting.',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle:
                  'Additional information to describe the Opportunity, such as possible products to sell or past purchases from the customer.',
              },
              {
                id: 'ownerid',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Dynamics 365 user or team assigned to this Opportunity. Use Connect Portal Workflow Settings to allow users to select an Owner (Team or User).',
                placeholder: '{{settings.ownerID}}',
              },
              {
                id: 'transactioncurrencyid',
                title: 'Transaction Currency',
                type: 'TEXTAREA',
                subtitle:
                  'Choose the local currency for the Opportunity. Use Connect Portal Workflow Settings to allow users to select a Transaction Currency, or the first available will be used as a default.',
                placeholder: '{{settings.transactionCurrency}}',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'name',
                title: 'Account Name',
                type: 'TEXTAREA',
                subtitle:
                  'The company or business name for the Account. This field is required.',
                required: true,
              },
              {
                id: 'ownerid',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Dynamics 365 user or team assigned to this Opportunity. Use Connect Portal Workflow Settings to allow users to select an Owner (Team or User).',
                placeholder: '{{settings.ownerID}}',
              },
              {
                id: 'websiteurl',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'telephone1',
                title: 'Business Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle:
                  "Additional information to describe the contact, such as an excerpt from the company's website.",
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'firstname',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'lastname',
                title: 'Last Name',
                type: 'TEXTAREA',
                subtitle: 'This field is required.',
                required: true,
              },
              {
                id: 'emailaddress1',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'telephone1',
                title: 'Business Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'jobtitle',
                title: 'Job Title',
                type: 'TEXTAREA',
                subtitle: 'Title of the lead, such as CEO or Vice President.',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle:
                  "Additional information to describe the contact, such as an excerpt from the company's website.",
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'subject',
                title: 'Topic',
                type: 'TEXTAREA',
                subtitle:
                  'Use a descriptive name, such as the expected order, company name, or marketing source list, to identify the lead.',
                required: true,
              },
              {
                id: 'firstname',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'lastname',
                title: 'Last Name',
                type: 'TEXTAREA',
                subtitle: 'This field is required.',
                required: true,
              },
              {
                id: 'emailaddress1',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'telephone1',
                title: 'Business Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'websiteurl',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'jobtitle',
                title: 'Job Title',
                type: 'TEXTAREA',
                subtitle: 'Title of the lead, such as CEO or Vice President.',
              },
              {
                id: 'salesstagecode',
                title: 'Process Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Process Stage.',
                placeholder: '{{settings.processStage}}',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle:
                  "Additional information to describe the lead, such as an excerpt from the company's website.",
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
          {
            value: 'Call',
            dependentInputs: [
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
                subtitle:
                  'A description of the objective or topic of the call.',
                required: true,
              },
              {
                id: 'from',
                title: 'From',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead, or user this call is from. Defaults to the currently authenticated user.',
              },
              {
                id: 'to',
                title: 'To',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead, or user this call is to. Use Connect Portal User Settings to allow users to specify a record.',
                placeholder: '{{settings.lead}}',
                required: true,
              },
              {
                id: 'directioncode',
                title: 'Direction',
                type: 'EDITABLE_ENUM',
                subtitle: 'Defaults to Outgoing.',
                values: ['1', '0'],
              },
              {
                id: 'regardingobjectid',
                title: 'Regarding',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead or opportunity this call is about.',
                placeholder: '{{settings.opportunity}}',
              },
              {
                id: 'phonenumber',
                title: 'Phone Number',
                type: 'TEXTAREA',
              },
              {
                id: 'actualdurationminutes',
                title: 'Duration',
                type: 'TEXTAREA',
                subtitle:
                  'Specify duration in minutes. Defaults to 30 minutes.',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'scheduledend',
                title: 'Due Date/Time',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO 8601 date formats.',
              },
              {
                id: 'prioritycode',
                title: 'Priority',
                type: 'EDITABLE_ENUM',
                subtitle: 'Defaults to Normal.',
                defaultValue: '1',
                values: ['0', '1', '2'],
              },
            ],
          },
          {
            value: 'Meeting',
            dependentInputs: [
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
                subtitle:
                  'A description of the objective or topic of the meeting.',
                required: true,
              },
              {
                id: 'scheduledstart',
                title: 'Start Date/Time',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO 8601 date formats.',
                required: true,
              },
              {
                id: 'scheduledend',
                title: 'End Date/Time',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO 8601 date formats.',
                required: true,
              },
              {
                id: 'requiredattendees',
                title: 'Attendees',
                type: 'CODE',
                subtitle:
                  'Add accounts, contacts, leads, or users to this meeting. Specify as an array of IDs.',
                placeholder:
                  '\n  [\n    "{{userSettings.accountId}}",\n    "/systemusers(1d9e0de7-4fb2-ec11-9840-0022480b1bdf)"\n  ]',
              },
              {
                id: 'regardingobjectid',
                title: 'Regarding',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead or opportunity this meeting is about.',
                placeholder: '{{settings.opportunity}}',
              },
              {
                id: 'location',
                title: 'Location',
                type: 'TEXTAREA',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'prioritycode',
                title: 'Priority',
                type: 'EDITABLE_ENUM',
                subtitle: 'Defaults to Normal.',
                defaultValue: '1',
                values: ['0', '1', '2'],
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
                subtitle: 'A title for the task.',
                required: true,
              },
              {
                id: 'regardingobjectid',
                title: 'Regarding',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead or opportunity this task is about.',
                placeholder: '{{settings.opportunity}}',
              },
              {
                id: 'scheduledend',
                title: 'Due Date/Time',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO 8601 date formats.',
              },
              {
                id: 'ownerid',
                title: 'Owner',
                type: 'TEXTAREA',
                subtitle:
                  'Specify a user to own this task. Defaults to the currently authenticated user.',
                placeholder: '{{settings.user}}',
              },
              {
                id: 'prioritycode',
                title: 'Priority',
                type: 'EDITABLE_ENUM',
                subtitle: 'Defaults to Normal.',
                defaultValue: '1',
                values: ['0', '1', '2'],
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: 'Note',
            dependentInputs: [
              {
                id: 'objectid',
                title: 'For',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead, or opportunity this note is attached to.',
                placeholder: '{{settings.opportunity}}',
                required: true,
              },
              {
                id: 'subject',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle: 'A title for the note.',
                required: true,
              },
              {
                id: 'notetext',
                title: 'Description',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: '<specified by entity property>',
            dependentInputs: [
              {
                id: 'customMdEntityName',
                title: 'Entity Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Entity Set Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: 'entitysetname',
                required: true,
              },
              {
                id: 'additionalFields',
                title: 'Record Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Field Mapping to specify which fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Update record',
    name: 'MICROSOFT_DYNAMICS_UPDATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'mdEntity',
        title: 'Choose an entity',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Opportunity ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Opportunity to update.',
                required: true,
              },
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                subtitle:
                  'A subject or descriptive name, such as the expected order or company name, for the Opportunity.',
              },
              {
                id: 'customerid',
                title: 'Potential Customer ID',
                type: 'TEXTAREA',
                subtitle:
                  'Select an account or contact to associate with this Opportunity.',
                placeholder: '{{settings.customerID}}',
              },
              {
                id: 'salesstagecode',
                title: 'Process Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Process Stage.',
                placeholder: '{{settings.processStage}}',
              },
              {
                id: 'estimatedclosedate',
                title: 'Estimated Close Date',
                type: 'TEXTAREA',
                subtitle: 'The expected closing date of the Opportunity.',
              },
              {
                id: 'actualclosedate',
                title: 'Actual Close Date',
                type: 'TEXTAREA',
                subtitle:
                  'The date and time when the Opportunity was closed or canceled.',
              },
              {
                id: 'estimatedvalue',
                title: 'Est. Revenue',
                type: 'TEXTAREA',
                subtitle:
                  'The estimated revenue amount to indicate the potential sale or value of the opportunity for revenue forecasting.',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle:
                  'Additional information to describe the Opportunity, such as possible products to sell or past purchases from the customer.',
              },
              {
                id: 'ownerid',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Dynamics 365 user or team assigned to this Opportunity. Use Connect Portal Workflow Settings to allow users to select an Owner (Team or User).',
                placeholder: '{{settings.ownerID}}',
              },
              {
                id: 'transactioncurrencyid',
                title: 'Transaction Currency',
                type: 'TEXTAREA',
                subtitle:
                  'Choose the local currency for the Opportunity. Use Connect Portal Workflow Settings to allow users to select a Transaction Currency, or the first available will be used as a default.',
                placeholder: '{{settings.transactionCurrency}}',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Account ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Account to update.',
                required: true,
              },
              {
                id: 'name',
                title: 'Account Name',
                type: 'TEXTAREA',
                subtitle: 'The company or business name for the Account.',
              },
              {
                id: 'ownerid',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Dynamics 365 user or team assigned to this Opportunity. Use Connect Portal Workflow Settings to allow users to select an Owner (Team or User).',
                placeholder: '{{settings.ownerID}}',
              },
              {
                id: 'websiteurl',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'telephone1',
                title: 'Business Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle:
                  "Additional information to describe the contact, such as an excerpt from the company's website.",
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Contact to update.',
                required: true,
              },
              {
                id: 'firstname',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'lastname',
                title: 'Last Name',
                type: 'TEXTAREA',
                subtitle: 'This field is required.',
              },
              {
                id: 'emailaddress1',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'telephone1',
                title: 'Business Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'jobtitle',
                title: 'Job Title',
                type: 'TEXTAREA',
                subtitle: 'Title of the lead, such as CEO or Vice President.',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle:
                  "Additional information to describe the contact, such as an excerpt from the company's website.",
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Lead to update.',
                required: true,
              },
              {
                id: 'subject',
                title: 'Topic',
                type: 'TEXTAREA',
                subtitle:
                  'Use a descriptive name, such as the expected order, company name, or marketing source list, to identify the lead.',
              },
              {
                id: 'firstname',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'lastname',
                title: 'Last Name',
                type: 'TEXTAREA',
                subtitle: 'This field is required.',
              },
              {
                id: 'emailaddress1',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'telephone1',
                title: 'Business Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'websiteurl',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'jobtitle',
                title: 'Job Title',
                type: 'TEXTAREA',
                subtitle: 'Title of the lead, such as CEO or Vice President.',
              },
              {
                id: 'salesstagecode',
                title: 'Process Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Process Stage.',
                placeholder: '{{settings.processStage}}',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle:
                  "Additional information to describe the lead, such as an excerpt from the company's website.",
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
          {
            value: 'Call',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Call ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Call to update.',
                required: true,
              },
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
                subtitle:
                  'A description of the objective or topic of the call.',
              },
              {
                id: 'from',
                title: 'From',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead, or user this call is from. Defaults to the currently authenticated user.',
              },
              {
                id: 'to',
                title: 'To',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead, or user this call is to. Use Connect Portal User Settings to allow users to specify a record.',
                placeholder: '{{settings.lead}}',
              },
              {
                id: 'directioncode',
                title: 'Direction',
                type: 'EDITABLE_ENUM',
                subtitle: 'Defaults to Outgoing.',
                values: ['1', '0'],
              },
              {
                id: 'regardingobjectid',
                title: 'Regarding',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead or opportunity this call is about.',
                placeholder: '{{settings.opportunity}}',
              },
              {
                id: 'phonenumber',
                title: 'Phone Number',
                type: 'TEXTAREA',
              },
              {
                id: 'actualdurationminutes',
                title: 'Duration',
                type: 'TEXTAREA',
                subtitle:
                  'Specify duration in minutes. Defaults to 30 minutes.',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'scheduledend',
                title: 'Due Date/Time',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO 8601 date formats.',
              },
              {
                id: 'prioritycode',
                title: 'Priority',
                type: 'EDITABLE_ENUM',
                subtitle: 'Defaults to Normal.',
                defaultValue: '1',
                values: ['0', '1', '2'],
              },
            ],
          },
          {
            value: 'Meeting',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Meeting ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Meeting to update.',
                required: true,
              },
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
                subtitle:
                  'A description of the objective or topic of the meeting.',
              },
              {
                id: 'scheduledstart',
                title: 'Start Date/Time',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO 8601 date formats.',
              },
              {
                id: 'scheduledend',
                title: 'End Date/Time',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO 8601 date formats.',
              },
              {
                id: 'requiredattendees',
                title: 'Attendees',
                type: 'CODE',
                subtitle:
                  'Add accounts, contacts, leads, or users to this meeting. Specify as an array of IDs.',
                placeholder:
                  '\n  [\n    "{{userSettings.accountId}}",\n    "/systemusers(1d9e0de7-4fb2-ec11-9840-0022480b1bdf)"\n  ]',
              },
              {
                id: 'regardingobjectid',
                title: 'Regarding',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead or opportunity this meeting is about.',
                placeholder: '{{settings.opportunity}}',
              },
              {
                id: 'location',
                title: 'Location',
                type: 'TEXTAREA',
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'prioritycode',
                title: 'Priority',
                type: 'EDITABLE_ENUM',
                subtitle: 'Defaults to Normal.',
                defaultValue: '1',
                values: ['0', '1', '2'],
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Task ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Task to update.',
                required: true,
              },
              {
                id: 'subject',
                title: 'Subject',
                type: 'TEXTAREA',
                subtitle: 'A title for the task.',
              },
              {
                id: 'regardingobjectid',
                title: 'Regarding',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead or opportunity this task is about.',
                placeholder: '{{settings.opportunity}}',
              },
              {
                id: 'scheduledend',
                title: 'Due Date/Time',
                type: 'TEXTAREA',
                subtitle: 'Accepts Unix timestamp or ISO 8601 date formats.',
              },
              {
                id: 'ownerid',
                title: 'Owner',
                type: 'TEXTAREA',
                subtitle:
                  'Specify a user to own this task. Defaults to the currently authenticated user.',
                placeholder: '{{settings.user}}',
              },
              {
                id: 'prioritycode',
                title: 'Priority',
                type: 'EDITABLE_ENUM',
                subtitle: 'Defaults to Normal.',
                defaultValue: '1',
                values: ['0', '1', '2'],
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: 'Note',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Note ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Note to update.',
                required: true,
              },
              {
                id: 'objectid',
                title: 'For',
                type: 'TEXTAREA',
                subtitle:
                  'The account, contact, lead, or opportunity this note is attached to.',
                placeholder: '{{settings.opportunity}}',
              },
              {
                id: 'subject',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle: 'A title for the note.',
              },
              {
                id: 'notetext',
                title: 'Description',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: '<specified by entity property>',
            dependentInputs: [
              {
                id: 'customMdEntityName',
                title: 'Entity Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Entity Set Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: 'entitysetname',
                required: true,
              },
              {
                id: 'entityId',
                title: 'Entity ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the entity to update.',
                required: true,
              },
              {
                id: 'additionalFields',
                title: 'Record Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Field Mapping to specify which fields to update.',
                placeholder:
                  '{\n    "address2_city": "Los Angeles",\n    "address2_stateorprovince": "CA",\n    "address2_country": "United States",\n    "address2_line1": "1700 Sawtelle Blvd",\n    "primarycontactid@odata.bind": "/contacts(00000000-\n      0000-0000-0000-000000000001)",\n}',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get record by ID',
    name: 'MICROSOFT_DYNAMICS_GET_RECORD_BY_ID',
    output: [],
    inputs: [
      {
        id: 'mdEntity',
        title: 'Choose an entity',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Opportunity ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Opportunity to fetch.',
                required: true,
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Account ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Account to fetch.',
                required: true,
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Contact to fetch.',
                required: true,
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Lead to fetch.',
                required: true,
              },
            ],
          },
          {
            value: 'Call',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Call ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Call to fetch.',
                required: true,
              },
            ],
          },
          {
            value: 'Meeting',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Meeting ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Meeting to fetch.',
                required: true,
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Task ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Task to fetch.',
                required: true,
              },
            ],
          },
          {
            value: 'Note',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Note ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Note to fetch.',
                required: true,
              },
            ],
          },
          {
            value: '<specified by entity property>',
            dependentInputs: [
              {
                id: 'customMdEntityName',
                title: 'Entity Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Entity Set Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: 'entitysetname',
                required: true,
              },
              {
                id: 'entityId',
                title: 'Entity ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the entity to fetch.',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Search records',
    name: 'MICROSOFT_DYNAMICS_SEARCH_RECORDS',
    output: [],
    inputs: [
      {
        id: 'mdEntity',
        title: 'Choose an entity',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'name',
                  'customerid',
                  'salesstagecode',
                  'estimatedclosedate',
                  'actualclosedate',
                  'estimatedvalue',
                  'description',
                  '_ownerid_value',
                  'transactioncurrencyid',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'name',
                  '_ownerid_value',
                  'websiteurl',
                  'telephone1',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'firstname',
                  'lastname',
                  'emailaddress1',
                  'telephone1',
                  'jobtitle',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'firstname',
                  'lastname',
                  'emailaddress1',
                  'telephone1',
                  'websiteurl',
                  'jobtitle',
                  'salesstagecode',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Call',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'directioncode',
                  '_regardingobjectid_value',
                  'phonenumber',
                  'actualdurationminutes',
                  'description',
                  'scheduledend',
                  'prioritycode',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Meeting',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'scheduledstart',
                  'scheduledend',
                  '_regardingobjectid_value',
                  'location',
                  'description',
                  'prioritycode',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  '_regardingobjectid_value',
                  'scheduledend',
                  '_ownerid_value',
                  'prioritycode',
                  'description',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: 'Note',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: ['_objectid_value', 'subject', 'notetext'],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
          {
            value: '<specified by entity property>',
            dependentInputs: [
              {
                id: 'customMdEntityName',
                title: 'Entity Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Entity Set Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: 'entitysetname',
                required: true,
              },
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringContains',
                  '$numberLessThan',
                  '$dateTimeBefore',
                  '$numberEquals',
                  '$dateTimeEquals',
                  '$numberGreaterThan',
                  '$dateTimeAfter',
                  '$numberDoesNotEqual',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                  '$booleanTrue',
                  '$booleanFalse',
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Delete Record by ID',
    name: 'MICROSOFT_DYNAMICS_DELETE_RECORD',
    output: [],
    inputs: [
      {
        id: 'mdEntity',
        title: 'Choose an entity',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Opportunity ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Opportunity to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Account ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Account to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Contact to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Lead to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'Call',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Call ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Call to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'Meeting',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Meeting ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Meeting to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Task ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Task to delete.',
                required: true,
              },
            ],
          },
          {
            value: 'Note',
            dependentInputs: [
              {
                id: 'entityId',
                title: 'Note ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the Note to delete.',
                required: true,
              },
            ],
          },
          {
            value: '<specified by entity property>',
            dependentInputs: [
              {
                id: 'customMdEntityName',
                title: 'Entity Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Entity Set Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: 'entitysetname',
                required: true,
              },
              {
                id: 'entityId',
                title: 'Entity ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the entity to delete.',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
];

/**
 * integration class for microsoftDynamics
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'microsoftDynamics',

  /**
   * action step methods
   */
  actions: {
    /**
     * Create record
     * method for MICROSOFT_DYNAMICS_CREATE_RECORD
     */
    microsoftDynamicsCreateRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'MICROSOFT_DYNAMICS_CREATE_RECORD',
        actionType: 'microsoftDynamics',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update record
     * method for MICROSOFT_DYNAMICS_UPDATE_RECORD
     */
    microsoftDynamicsUpdateRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'MICROSOFT_DYNAMICS_UPDATE_RECORD',
        actionType: 'microsoftDynamics',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get record by ID
     * method for MICROSOFT_DYNAMICS_GET_RECORD_BY_ID
     */
    microsoftDynamicsGetRecordById: (inputs, options) => {
      return new ActionStep({
        intent: 'MICROSOFT_DYNAMICS_GET_RECORD_BY_ID',
        actionType: 'microsoftDynamics',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Search records
     * method for MICROSOFT_DYNAMICS_SEARCH_RECORDS
     */
    microsoftDynamicsSearchRecords: (inputs, options) => {
      return new ActionStep({
        intent: 'MICROSOFT_DYNAMICS_SEARCH_RECORDS',
        actionType: 'microsoftDynamics',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Delete Record by ID
     * method for MICROSOFT_DYNAMICS_DELETE_RECORD
     */
    microsoftDynamicsDeleteRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'MICROSOFT_DYNAMICS_DELETE_RECORD',
        actionType: 'microsoftDynamics',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {
    /**
     * New Record
     * method for MICROSOFT_DYNAMICS_TRIGGER_RECORD_CREATED
     */
    microsoftDynamicsTriggerRecordCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'MICROSOFT_DYNAMICS_TRIGGER_RECORD_CREATED',
        actionType: 'microsoftDynamics',
        parameters: inputs,
      });
    },
    /**
     * Record Updated
     * method for MICROSOFT_DYNAMICS_TRIGGER_RECORD_UPDATED
     */
    microsoftDynamicsTriggerRecordUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'MICROSOFT_DYNAMICS_TRIGGER_RECORD_UPDATED',
        actionType: 'microsoftDynamics',
        parameters: inputs,
      });
    },
  },
};
