// @useparagon/core/dist/steps/library/request/request.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLIENT_CREDENTIAL_REQUEST_TYPE = exports.RequestAuthorizationType = exports.RequestBodyType = exports.HttpMethod = void 0;
var HttpMethod;
(function (HttpMethod) {
    HttpMethod["GET"] = "GET";
    HttpMethod["POST"] = "POST";
    HttpMethod["PUT"] = "PUT";
    HttpMethod["PATCH"] = "PATCH";
    HttpMethod["DELETE"] = "DELETE";
})(HttpMethod || (exports.HttpMethod = HttpMethod = {}));
var RequestBodyType;
(function (RequestBodyType) {
    RequestBodyType["JSON"] = "json";
    RequestBodyType["FORM_DATA"] = "form-data";
    RequestBodyType["X_WWW_FORM_URLENCODED"] = "x-www-form-urlencoded";
    RequestBodyType["XML"] = "xml";
    RequestBodyType["RAW"] = "raw";
})(RequestBodyType || (exports.RequestBodyType = RequestBodyType = {}));
var RequestAuthorizationType;
(function (RequestAuthorizationType) {
    RequestAuthorizationType["NONE"] = "none";
    RequestAuthorizationType["BEARER_TOKEN"] = "bearer";
    RequestAuthorizationType["BASIC"] = "basic";
    RequestAuthorizationType["QUERY_PARAMS"] = "query_params";
    RequestAuthorizationType["AUTH_HEADER"] = "auth_header";
    RequestAuthorizationType["CLIENT_CREDENTIAL"] = "oauth-client-credential";
    RequestAuthorizationType["REFRESH_TOKEN"] = "oauth-refresh-token";
})(RequestAuthorizationType || (exports.RequestAuthorizationType = RequestAuthorizationType = {}));
var CLIENT_CREDENTIAL_REQUEST_TYPE;
(function (CLIENT_CREDENTIAL_REQUEST_TYPE) {
    CLIENT_CREDENTIAL_REQUEST_TYPE["BodyParam"] = "body-param";
    CLIENT_CREDENTIAL_REQUEST_TYPE["BasicAuthHeader"] = "basic-auth-header";
})(CLIENT_CREDENTIAL_REQUEST_TYPE || (exports.CLIENT_CREDENTIAL_REQUEST_TYPE = CLIENT_CREDENTIAL_REQUEST_TYPE = {}));
