// @useparagon/integrations/close/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for close
 */
exports.inputs = [
  {
    id: 'users',
    title: 'Users',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'leads',
    title: 'Leads',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'getStatusCombo',
    title: 'Status',
    type: 'COMBO_INPUT_DATA_SOURCE',
    mainInputSource: {
      id: 'recordType',
      title: 'Record Type',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    dependentInputSource: {
      id: 'status',
      title: 'Status',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
  {
    id: 'customFieldCombo',
    title: 'Custom Fields',
    type: 'COMBO_INPUT_DATA_SOURCE',
    mainInputSource: {
      id: 'recordTypeForCustomField',
      title: 'Record Type',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    dependentInputSource: {
      id: 'customField',
      title: 'Custom Fields',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
  {
    id: 'mapCloseFields',
    title: 'Field Mapping',
    type: 'FIELD_MAPPER_DATA_SOURCE',
    fieldSource: {
      id: 'customField',
      title: 'Custom Fields',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    recordSource: {
      id: 'recordTypeForCustomField',
      title: 'Record Type',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
];

/**
 * triggers for close
 */
exports.triggers = [
  {
    title: 'Record Created',
    name: 'CLOSE_TRIGGER_RECORD_CREATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          'lead',
          'opportunity',
          'contact',
          {
            value: 'task',
            dependentInputs: [
              {
                id: 'recordSubTypes',
                title: 'Record Subtype',
                type: 'TEXTAREA',
                placeholder: '["lead", "incoming_email", "email_followup"]',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'triggerFilterKey',
        title: 'Filter Records',
        type: 'CONDITIONAL',
        subtitle: 'Search for the records that match specified filters.',
        placeholder: 'field',
        supportedOperators: [
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringContains',
          '$doesNotExist',
          '$exists',
        ],
      },
    ],
  },
  {
    title: 'Record Updated',
    name: 'CLOSE_TRIGGER_RECORD_UPDATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          'lead',
          'opportunity',
          'contact',
          {
            value: 'task',
            dependentInputs: [
              {
                id: 'recordSubTypes',
                title: 'Record Subtype',
                type: 'TEXTAREA',
                placeholder: '["lead", "incoming_email", "email_followup"]',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'triggerFilterKey',
        title: 'Filter Records',
        type: 'CONDITIONAL',
        subtitle: 'Search for the records that match specified filters.',
        placeholder: 'field',
        supportedOperators: [
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringContains',
          '$doesNotExist',
          '$exists',
        ],
      },
    ],
  },
  {
    title: 'Record Deleted',
    name: 'CLOSE_TRIGGER_RECORD_DELETED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          'lead',
          'opportunity',
          'contact',
          {
            value: 'task',
            dependentInputs: [
              {
                id: 'recordSubTypes',
                title: 'Record Subtype',
                type: 'TEXTAREA',
                placeholder: '["lead", "incoming_email", "email_followup"]',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'triggerFilterKey',
        title: 'Filter Records',
        type: 'CONDITIONAL',
        subtitle: 'Search for the records that match specified filters.',
        placeholder: 'field',
        supportedOperators: [
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringContains',
          '$doesNotExist',
          '$exists',
        ],
      },
    ],
  },
];

/**
 * intents for close
 */
exports.intents = [
  {
    title: 'Create Record',
    name: 'CLOSE_CREATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'lead',
            dependentInputs: [
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                subtitle: 'Specify the name of the lead.',
              },
              {
                id: 'leadDescription',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'Specify a description for this lead.',
              },
              {
                id: 'leadContact',
                title: 'Contacts',
                type: 'CODE',
                subtitle:
                  'Specify contact details as JSON to attach to this lead.',
                placeholder:
                  '[\n  {\n    "name": "Gob",\n    "title": "Sr. Vice President",\n    "emails": [\n      {\n        "type": "office",\n        "email": "gob@example.com"\n      }\n    ],\n    "phones": [\n      {\n        "type": "office",\n        "phone": "8004445555"\n      }\n    ]\n  }\n]',
              },
              {
                id: 'leadAddress',
                title: 'Addresses',
                type: 'CODE',
                subtitle:
                  'Specify address details as JSON to attach to this lead.',
                placeholder:
                  '[\n  {\n    "label": "business",\n    "address_1": "747 Howard St",\n    "address_2": "Room 3",\n    "city": "San Francisco",\n    "state": "CA",\n    "zipcode": "94103",\n    "country":"US",\n  }\n]',
              },
              {
                id: 'status',
                title: 'Status',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select Status use for this lead.',
              },
              {
                id: 'leadAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                placeholder:
                  '{\n  "confidence": 90,\n  "value_period": "monthly",\n  "custom.cf_j0P7kHmgFTZZnYBFtyPSZ3uQw4dpW8xKcW7Krps8atj": "https://www.example.com/contract.pdf"\n}',
              },
            ],
          },
          {
            value: 'opportunity',
            dependentInputs: [
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle:
                  'Specify the ID of the Lead this opportunity is associated with.',
                required: true,
              },
              {
                id: 'value',
                title: 'Value',
                type: 'TEXTAREA',
                subtitle: 'Specify a value for this opportunity.',
                placeholder: '500',
              },
              {
                id: 'note',
                title: 'Note',
                type: 'TEXTAREA',
                subtitle: 'Specify a note to attach to this opportunity',
              },
              {
                id: 'opportunityStatus',
                title: 'Status',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select Status use for this opportunity.',
              },
              {
                id: 'confidence',
                title: 'Confidence',
                type: 'TEXTAREA',
              },
              {
                id: 'opportunityAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                placeholder:
                  '{\n  "value_period": "monthly",\n  "custom.cf_j0P7kHmgFTZZnYBFtyPSZ3uQw4dpW8xKcW7Krps8atj": "https://www.example.com/contract.pdf"\n}',
              },
            ],
          },
          {
            value: 'contact',
            dependentInputs: [
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle:
                  'Specify ID of the Lead this contact is associated with.',
                required: true,
              },
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                subtitle: 'Specify a name for this contact.',
              },
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle: 'Specify a title for this contact.',
                placeholder: 'President',
              },
              {
                id: 'phone',
                title: 'Phones',
                type: 'CODE',
                placeholder: '[\n  {"phone":"9045551234","type":"mobile"}\n]',
              },
              {
                id: 'email',
                title: 'Emails',
                type: 'CODE',
                placeholder:
                  '[\n  {"email":"john@example.com","type":"office"}\n]',
              },
              {
                id: 'urls',
                title: 'Urls',
                type: 'CODE',
                placeholder:
                  '[\n  {"url":"http://twitter.com/google/","type":"url"}\n]',
              },
              {
                id: 'contactAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                placeholder:
                  '{\n  "custom.cf_FSYEbxYJFsnY9tN1OTAPIF33j7Sw5Lb7Eawll7JzoNh": "Account Executive",\n}',
              },
            ],
          },
          {
            value: 'task',
            dependentInputs: [
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle:
                  'Specify ID of the Lead this task is associated with.',
                required: true,
              },
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle: 'Specify a title for this contact.',
                placeholder: 'Connect with Account Manager',
                required: true,
              },
              {
                id: 'assignedTo',
                title: 'Assigned To',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select user to assign this task to.',
                placeholder: '{{settings.user}}',
              },
              {
                id: 'date',
                title: 'Date',
                type: 'TEXTAREA',
                placeholder: '2013-02-06',
              },
              {
                id: 'isComplete',
                title: 'Mark task as completed',
                type: 'BOOLEAN_INPUT',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Update Record',
    name: 'CLOSE_UPDATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'lead',
            dependentInputs: [
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the Id of Lead to update.',
                required: true,
              },
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                subtitle: 'Specify the name of the lead.',
              },
              {
                id: 'leadDescription',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'Specify a description for this lead.',
              },
              {
                id: 'leadContact',
                title: 'Contacts',
                type: 'CODE',
                subtitle:
                  'Specify contact details as JSON to attach to this lead.',
                placeholder:
                  '[\n  {\n    "name": "Gob",\n    "title": "Sr. Vice President",\n    "emails": [\n      {\n        "type": "office",\n        "email": "gob@example.com"\n      }\n    ],\n    "phones": [\n      {\n        "type": "office",\n        "phone": "8004445555"\n      }\n    ]\n  }\n]',
              },
              {
                id: 'leadAddress',
                title: 'Addresses',
                type: 'CODE',
                subtitle:
                  'Specify address details as JSON to attach to this lead.',
                placeholder:
                  '[\n  {\n    "label": "business",\n    "address_1": "747 Howard St",\n    "address_2": "Room 3",\n    "city": "San Francisco",\n    "state": "CA",\n    "zipcode": "94103",\n    "country":"US",\n  }\n]',
              },
              {
                id: 'status',
                title: 'Status',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select Status use for this lead.',
              },
              {
                id: 'leadAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                placeholder:
                  '{\n  "confidence": 90,\n  "value_period": "monthly",\n  "custom.cf_j0P7kHmgFTZZnYBFtyPSZ3uQw4dpW8xKcW7Krps8atj": "https://www.example.com/contract.pdf"\n}',
              },
            ],
          },
          {
            value: 'opportunity',
            dependentInputs: [
              {
                id: 'opportunityId',
                title: 'Opportunity ID',
                type: 'TEXTAREA',
                subtitle: 'Specify Id of opportunity to update.',
                required: true,
              },
              {
                id: 'value',
                title: 'Value',
                type: 'TEXTAREA',
                subtitle: 'Specify a value for this opportunity.',
                placeholder: '500',
              },
              {
                id: 'note',
                title: 'Note',
                type: 'TEXTAREA',
                subtitle: 'Specify a note to attach to this opportunity',
              },
              {
                id: 'opportunityStatus',
                title: 'Status',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select Status use for this opportunity.',
              },
              {
                id: 'confidence',
                title: 'Confidence',
                type: 'TEXTAREA',
              },
              {
                id: 'opportunityAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                placeholder:
                  '{\n  "value_period": "monthly",\n  "custom.cf_j0P7kHmgFTZZnYBFtyPSZ3uQw4dpW8xKcW7Krps8atj": "https://www.example.com/contract.pdf"\n}',
              },
            ],
          },
          {
            value: 'contact',
            dependentInputs: [
              {
                id: 'contactId',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the of Contact to update',
                required: true,
              },
              {
                id: 'name',
                title: 'Name',
                type: 'TEXTAREA',
                subtitle: 'Specify a name for this contact.',
              },
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle: 'Specify a title for this contact.',
                placeholder: 'President',
              },
              {
                id: 'phone',
                title: 'Phones',
                type: 'CODE',
                placeholder: '[\n  {"phone":"9045551234","type":"mobile"}\n]',
              },
              {
                id: 'email',
                title: 'Emails',
                type: 'CODE',
                placeholder:
                  '[\n  {"email":"john@example.com","type":"office"}\n]',
              },
              {
                id: 'urls',
                title: 'Urls',
                type: 'CODE',
                placeholder:
                  '[\n  {"url":"http://twitter.com/google/","type":"url"}\n]',
              },
              {
                id: 'contactAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                placeholder:
                  '{\n  "custom.cf_FSYEbxYJFsnY9tN1OTAPIF33j7Sw5Lb7Eawll7JzoNh": "Account Executive",\n}',
              },
            ],
          },
          {
            value: 'task',
            dependentInputs: [
              {
                id: 'taskId',
                title: 'Task ID',
                type: 'TEXTAREA',
                subtitle: 'Specify Id of the task to update.',
                required: true,
              },
              {
                id: 'title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle: 'Specify a title for this contact.',
                placeholder: 'Connect with Account Manager',
                required: true,
              },
              {
                id: 'assignedTo',
                title: 'Assigned To',
                type: 'TEXTAREA',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select user to assign this task to.',
                placeholder: '{{settings.user}}',
              },
              {
                id: 'date',
                title: 'Date',
                type: 'TEXTAREA',
                placeholder: '2013-02-06',
              },
              {
                id: 'isComplete',
                title: 'Mark task as completed',
                type: 'BOOLEAN_INPUT',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get Record by ID',
    name: 'CLOSE_GET_RECORD_BY_ID',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'lead',
            dependentInputs: [
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the ID of lead to fetch',
                required: true,
              },
            ],
          },
          {
            value: 'opportunity',
            dependentInputs: [
              {
                id: 'opportunityId',
                title: 'Opportunity ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the ID of opportunity to fetch',
                required: true,
              },
            ],
          },
          {
            value: 'contact',
            dependentInputs: [
              {
                id: 'contactId',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the ID of contact to fetch',
                required: true,
              },
            ],
          },
          {
            value: 'task',
            dependentInputs: [
              {
                id: 'taskId',
                title: 'Task ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the ID of task to fetch',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Search Record',
    name: 'CLOSE_SEARCH_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          'lead',
          {
            value: 'opportunity',
            dependentInputs: [
              {
                id: 'opportunityFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle:
                  'Search for opportunity that match specified filters.',
                placeholder: 'field',
                supportedKeys: [
                  'id',
                  '_type',
                  'lead_id',
                  'user_id',
                  'value_period',
                  'status_label',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$stringIsIn',
                  '$stringIsNotIn',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                ],
              },
            ],
          },
          'contact',
          {
            value: 'task',
            dependentInputs: [
              {
                id: 'taskFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for task that match specified filters.',
                placeholder: 'field',
                supportedKeys: [
                  'id',
                  '_type',
                  'lead_id',
                  'membership_id',
                  'assigned_to',
                  'view',
                ],
                supportedOperators: [
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$stringIsIn',
                  '$stringIsNotIn',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberGreaterThanOrEqualTo',
                  '$numberLessThanOrEqualTo',
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Delete Record',
    name: 'CLOSE_DELETE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'lead',
            dependentInputs: [
              {
                id: 'leadId',
                title: 'Lead ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the ID of lead to delete',
                required: true,
              },
            ],
          },
          {
            value: 'opportunity',
            dependentInputs: [
              {
                id: 'opportunityId',
                title: 'Opportunity ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the ID of opportunity to delete',
                required: true,
              },
            ],
          },
          {
            value: 'contact',
            dependentInputs: [
              {
                id: 'contactId',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the ID of contact to delete',
                required: true,
              },
            ],
          },
          {
            value: 'task',
            dependentInputs: [
              {
                id: 'taskId',
                title: 'Task ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the ID of task to delete',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Search records by JSON filter',
    name: 'CLOSE_ADVANCED_FILTERING',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: ['lead', 'opportunity', 'contact', 'task'],
      },
      {
        id: 'jsonFilter',
        title: 'Json Filter',
        type: 'CODE',
        placeholder:
          '{\n  "query":{ \n    "queries": [\n      {\n        "object_type": "contact",\n        "type": "object_type"\n      },\n      {\n        "type": "field_condition",\n        "field": {\n          "object_type": "contact",\n          "type": "regular_field",\n          "field_name": "title"\n        },\n        "condition": {\n          "type": "text",\n          "mode": "full_words",\n          "value": "CEO"\n        }\n      }\n    ],\n    "type": "and",\n    },\n  "_fields": {\n    "contact": ["id", "name", "title"]\n  }\n}',
        required: true,
      },
    ],
  },
];

/**
 * integration class for close
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'close',

  /**
   * action step methods
   */
  actions: {
    /**
     * Create Record
     * method for CLOSE_CREATE_RECORD
     */
    createRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'CLOSE_CREATE_RECORD',
        actionType: 'close',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update Record
     * method for CLOSE_UPDATE_RECORD
     */
    updateRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'CLOSE_UPDATE_RECORD',
        actionType: 'close',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Record by ID
     * method for CLOSE_GET_RECORD_BY_ID
     */
    getRecordById: (inputs, options) => {
      return new ActionStep({
        intent: 'CLOSE_GET_RECORD_BY_ID',
        actionType: 'close',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Search Record
     * method for CLOSE_SEARCH_RECORD
     */
    searchRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'CLOSE_SEARCH_RECORD',
        actionType: 'close',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Delete Record
     * method for CLOSE_DELETE_RECORD
     */
    deleteRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'CLOSE_DELETE_RECORD',
        actionType: 'close',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Search records by JSON filter
     * method for CLOSE_ADVANCED_FILTERING
     */
    advancedFiltering: (inputs, options) => {
      return new ActionStep({
        intent: 'CLOSE_ADVANCED_FILTERING',
        actionType: 'close',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {
    /**
     * Record Created
     * method for CLOSE_TRIGGER_RECORD_CREATED
     */
    recordCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'CLOSE_TRIGGER_RECORD_CREATED',
        actionType: 'close',
        parameters: inputs,
      });
    },
    /**
     * Record Updated
     * method for CLOSE_TRIGGER_RECORD_UPDATED
     */
    recordUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'CLOSE_TRIGGER_RECORD_UPDATED',
        actionType: 'close',
        parameters: inputs,
      });
    },
    /**
     * Record Deleted
     * method for CLOSE_TRIGGER_RECORD_DELETED
     */
    recordDeleted: (inputs) => {
      return new ActionTriggerStep({
        intent: 'CLOSE_TRIGGER_RECORD_DELETED',
        actionType: 'close',
        parameters: inputs,
      });
    },
  },
};
