// @useparagon/integrations/outreach/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for outreach
 */
exports.inputs = [
  {
    id: 'opportunityStages',
    title: 'Opportunity Stage',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'sequences',
    title: 'Sequence',
    type: 'DYNAMIC_DATA_SOURCE',
  },
];

/**
 * triggers for outreach
 */
exports.triggers = [
  {
    title: 'New Account',
    name: 'OUTREACH_TRIGGER_ACCOUNT_CREATED',
    output: [],
    inputs: [],
  },
  {
    title: 'Account Updated',
    name: 'OUTREACH_TRIGGER_ACCOUNT_UPDATED',
    output: [],
    inputs: [],
  },
  {
    title: 'New Opportunity',
    name: 'OUTREACH_TRIGGER_OPPORTUNITY_CREATED',
    output: [],
    inputs: [],
  },
  {
    title: 'Opportunity Updated',
    name: 'OUTREACH_TRIGGER_OPPORTUNITY_UPDATED',
    output: [],
    inputs: [],
  },
  {
    title: 'New Prospect',
    name: 'OUTREACH_TRIGGER_PROSPECT_CREATED',
    output: [],
    inputs: [],
  },
  {
    title: 'Prospect Updated',
    name: 'OUTREACH_TRIGGER_PROSPECT_UPDATED',
    output: [],
    inputs: [],
  },
];

/**
 * intents for outreach
 */
exports.intents = [
  {
    title: 'Create Account',
    name: 'OUTREACH_CREATE_ACCOUNT',
    output: [],
    inputs: [
      {
        id: 'accountName',
        title: 'Account name',
        type: 'TEXTAREA',
        required: true,
      },
      {
        id: 'websiteUrl',
        title: 'Website URL',
        type: 'TEXTAREA',
      },
      {
        id: 'description',
        title: 'Description',
        type: 'TEXTAREA',
      },
      {
        id: 'companyType',
        title: 'Company type',
        type: 'TEXTAREA',
        subtitle:
          'A description of the company’s type (e.g. "Public Company").',
      },
      {
        id: 'industry',
        title: 'Industry',
        type: 'TEXTAREA',
      },
      {
        id: 'numberOfEmployees',
        title: 'Number of employees',
        type: 'TEXTAREA',
      },
      {
        id: 'customId',
        title: 'Custom ID',
        type: 'TEXTAREA',
        subtitle:
          'A custom ID for the account, often referencing an ID in an external system.',
      },
    ],
  },
  {
    title: 'Update Account',
    name: 'OUTREACH_UPDATE_ACCOUNT',
    output: [],
    inputs: [
      {
        id: 'accountId',
        title: 'Account ID',
        type: 'TEXTAREA',
        subtitle: 'The ID of the account to update.',
        required: true,
      },
      {
        id: 'accountName',
        title: 'Account name',
        type: 'TEXTAREA',
        required: true,
      },
      {
        id: 'websiteUrl',
        title: 'Website URL',
        type: 'TEXTAREA',
      },
      {
        id: 'description',
        title: 'Description',
        type: 'TEXTAREA',
      },
      {
        id: 'companyType',
        title: 'Company type',
        type: 'TEXTAREA',
        subtitle:
          'A description of the company’s type (e.g. "Public Company").',
      },
      {
        id: 'industry',
        title: 'Industry',
        type: 'TEXTAREA',
      },
      {
        id: 'numberOfEmployees',
        title: 'Number of employees',
        type: 'TEXTAREA',
      },
      {
        id: 'customId',
        title: 'Custom ID',
        type: 'TEXTAREA',
        subtitle:
          'A custom ID for the account, often referencing an ID in an external system.',
      },
    ],
  },
  {
    title: 'Get Accounts',
    name: 'OUTREACH_GET_ACCOUNTS',
    output: [],
    inputs: [
      {
        id: 'accountsFilterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Get Accounts that match the specified filters.',
        placeholder: 'field',
        supportedKeys: ['customId', 'domain', 'name', 'createdAt', 'updatedAt'],
        supportedOperators: [
          '$stringExactlyMatches',
          '$numberEquals',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberGreaterThanOrEqualTo',
          '$numberLessThanOrEqualTo',
          '$dateTimeEquals',
          '$dateTimeBefore',
          '$dateTimeAfter',
        ],
      },
    ],
  },
  {
    title: 'Get Account by ID',
    name: 'OUTREACH_GET_ACCOUNT_BY_ID',
    output: [],
    inputs: [
      {
        id: 'accountId',
        title: 'Account ID',
        type: 'TEXTAREA',
        required: true,
      },
    ],
  },
  {
    title: 'Create Opportunity',
    name: 'OUTREACH_CREATE_OPPORTUNITY',
    output: [],
    inputs: [
      {
        id: 'opportunityName',
        title: 'Opportunity Name',
        type: 'TEXTAREA',
        required: true,
      },
      {
        id: 'amount',
        title: 'Amount',
        type: 'TEXTAREA',
      },
      {
        id: 'accountId',
        title: 'Account ID',
        type: 'TEXTAREA',
        subtitle: 'The Account that the opportunity belongs to.',
      },
      {
        id: 'opportunityStage',
        title: 'Opportunity stage',
        type: 'TEXTAREA',
        subtitle:
          'Use Connect Portal Workflow Settings to allow users to select an Opportunity Stage.',
        placeholder: '{{settings.opportunityStage}}',
      },
      {
        id: 'closeDate',
        title: 'Close date',
        type: 'TEXTAREA',
        placeholder: 'YYYY-MM-DD',
      },
      {
        id: 'description',
        title: 'Description',
        type: 'TEXTAREA',
      },
    ],
  },
  {
    title: 'Update Opportunity',
    name: 'OUTREACH_UPDATE_OPPORTUNITY',
    output: [],
    inputs: [
      {
        id: 'opportunityId',
        title: 'Opportunity ID',
        type: 'TEXTAREA',
        subtitle: 'The ID of the opportunity to update.',
        required: true,
      },
      {
        id: 'opportunityName',
        title: 'Opportunity Name',
        type: 'TEXTAREA',
      },
      {
        id: 'amount',
        title: 'Amount',
        type: 'TEXTAREA',
      },
      {
        id: 'accountId',
        title: 'Account ID',
        type: 'TEXTAREA',
        subtitle: 'The Account that the opportunity belongs to.',
      },
      {
        id: 'opportunityStage',
        title: 'Opportunity stage',
        type: 'TEXTAREA',
        subtitle:
          'Use Connect Portal Workflow Settings to allow users to select an Opportunity Stage.',
        placeholder: '{{settings.opportunityStage}}',
      },
      {
        id: 'closeDate',
        title: 'Close date',
        type: 'TEXTAREA',
        placeholder: 'YYYY-MM-DD',
      },
      {
        id: 'description',
        title: 'Description',
        type: 'TEXTAREA',
      },
    ],
  },
  {
    title: 'Get Opportunities',
    name: 'OUTREACH_GET_OPPORTUNITIES',
    output: [],
    inputs: [
      {
        id: 'opportunityFilterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Get opportunities that match the specified filters.',
        placeholder: 'field',
        supportedKeys: [
          'probability',
          'riskScore',
          'name',
          'mapStatus',
          'createdAt',
          'updatedAt',
        ],
        supportedOperators: [
          '$stringExactlyMatches',
          '$numberEquals',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberGreaterThanOrEqualTo',
          '$numberLessThanOrEqualTo',
          '$dateTimeEquals',
          '$dateTimeBefore',
          '$dateTimeAfter',
        ],
      },
    ],
  },
  {
    title: 'Get Opportunity by ID',
    name: 'OUTREACH_GET_OPPORTUNITY_BY_ID',
    output: [],
    inputs: [
      {
        id: 'opportunityId',
        title: 'Opportunity ID',
        type: 'TEXTAREA',
        required: true,
      },
    ],
  },
  {
    title: 'Create Prospect',
    name: 'OUTREACH_CREATE_PROSPECT',
    output: [],
    inputs: [
      {
        id: 'email',
        title: 'Email address',
        type: 'TEXTAREA',
        required: true,
      },
      {
        id: 'firstName',
        title: 'First name',
        type: 'TEXTAREA',
      },
      {
        id: 'lastName',
        title: 'Last name',
        type: 'TEXTAREA',
      },
      {
        id: 'accountId',
        title: 'Account ID',
        type: 'TEXTAREA',
        subtitle: 'The Account that the prospect belongs to.',
      },
      {
        id: 'phone',
        title: 'Phone number',
        type: 'TEXTAREA',
      },
      {
        id: 'company',
        title: 'Company',
        type: 'TEXTAREA',
      },
      {
        id: 'occupation',
        title: 'Occupation',
        type: 'TEXTAREA',
      },
      {
        id: 'websiteUrl',
        title: 'Website URL',
        type: 'TEXTAREA',
      },
      {
        id: 'externalId',
        title: 'External ID',
        type: 'TEXTAREA',
        subtitle:
          'A custom ID for the prospect, often referencing an ID in an external system.',
      },
    ],
  },
  {
    title: 'Update Prospect',
    name: 'OUTREACH_UPDATE_PROSPECT',
    output: [],
    inputs: [
      {
        id: 'prospectId',
        title: 'Prospect ID',
        type: 'TEXTAREA',
        subtitle: 'The ID of the prospect to update.',
        required: true,
      },
      {
        id: 'email',
        title: 'Email address',
        type: 'TEXTAREA',
      },
      {
        id: 'firstName',
        title: 'First name',
        type: 'TEXTAREA',
      },
      {
        id: 'lastName',
        title: 'Last name',
        type: 'TEXTAREA',
      },
      {
        id: 'accountId',
        title: 'Account ID',
        type: 'TEXTAREA',
        subtitle: 'The Account that the prospect belongs to.',
      },
      {
        id: 'phone',
        title: 'Phone number',
        type: 'TEXTAREA',
      },
      {
        id: 'company',
        title: 'Company',
        type: 'TEXTAREA',
      },
      {
        id: 'occupation',
        title: 'Occupation',
        type: 'TEXTAREA',
      },
      {
        id: 'websiteUrl',
        title: 'Website URL',
        type: 'TEXTAREA',
      },
      {
        id: 'externalId',
        title: 'External ID',
        type: 'TEXTAREA',
        subtitle:
          'A custom ID for the prospect, often referencing an ID in an external system.',
      },
    ],
  },
  {
    title: 'Get Prospects',
    name: 'OUTREACH_GET_PROSPECTS',
    output: [],
    inputs: [
      {
        id: 'prospectsFilterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Get prospects that match the specified filters.',
        placeholder: 'field',
        supportedKeys: [
          'emails',
          'engagedScore',
          'lastName',
          'title',
          'mapStatus',
          'createdAt',
          'updatedAt',
          'engagedAt',
        ],
        supportedOperators: [
          '$stringExactlyMatches',
          '$numberEquals',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberGreaterThanOrEqualTo',
          '$numberLessThanOrEqualTo',
          '$dateTimeEquals',
          '$dateTimeBefore',
          '$dateTimeAfter',
        ],
      },
    ],
  },
  {
    title: 'Get Prospect by ID',
    name: 'OUTREACH_GET_PROSPECT_BY_ID',
    output: [],
    inputs: [
      {
        id: 'prospectId',
        title: 'Prospect ID',
        type: 'TEXTAREA',
        required: true,
      },
    ],
  },
  {
    title: 'Add Prospect to Sequence',
    name: 'OUTREACH_ADD_PROSPECT_TO_SEQUENCE',
    output: [],
    inputs: [
      {
        id: 'sequence',
        title: 'Sequence',
        type: 'TEXTAREA',
        subtitle:
          'The Sequence that the prospect should be added to. Use Connect Portal Workflow Settings to allow users to select a Sequence. ',
        placeholder: '{{settings.sequence}}',
        required: true,
      },
      {
        id: 'prospectId',
        title: 'Prospect ID',
        type: 'TEXTAREA',
        subtitle: 'ID of prospect to be added to the Sequence.',
        required: true,
      },
    ],
  },
];

/**
 * integration class for outreach
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'outreach',

  /**
   * action step methods
   */
  actions: {
    /**
     * Create Account
     * method for OUTREACH_CREATE_ACCOUNT
     */
    createAccount: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_CREATE_ACCOUNT',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update Account
     * method for OUTREACH_UPDATE_ACCOUNT
     */
    updateAccount: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_UPDATE_ACCOUNT',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Accounts
     * method for OUTREACH_GET_ACCOUNTS
     */
    getAccounts: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_GET_ACCOUNTS',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Account by ID
     * method for OUTREACH_GET_ACCOUNT_BY_ID
     */
    getAccountById: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_GET_ACCOUNT_BY_ID',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Create Opportunity
     * method for OUTREACH_CREATE_OPPORTUNITY
     */
    createOpportunity: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_CREATE_OPPORTUNITY',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update Opportunity
     * method for OUTREACH_UPDATE_OPPORTUNITY
     */
    updateOpportunity: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_UPDATE_OPPORTUNITY',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Opportunities
     * method for OUTREACH_GET_OPPORTUNITIES
     */
    getOpportunities: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_GET_OPPORTUNITIES',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Opportunity by ID
     * method for OUTREACH_GET_OPPORTUNITY_BY_ID
     */
    getOpportunityById: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_GET_OPPORTUNITY_BY_ID',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Create Prospect
     * method for OUTREACH_CREATE_PROSPECT
     */
    createProspect: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_CREATE_PROSPECT',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update Prospect
     * method for OUTREACH_UPDATE_PROSPECT
     */
    updateProspect: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_UPDATE_PROSPECT',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Prospects
     * method for OUTREACH_GET_PROSPECTS
     */
    getProspects: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_GET_PROSPECTS',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Prospect by ID
     * method for OUTREACH_GET_PROSPECT_BY_ID
     */
    getProspectById: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_GET_PROSPECT_BY_ID',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Add Prospect to Sequence
     * method for OUTREACH_ADD_PROSPECT_TO_SEQUENCE
     */
    addProspectToSequence: (inputs, options) => {
      return new ActionStep({
        intent: 'OUTREACH_ADD_PROSPECT_TO_SEQUENCE',
        actionType: 'outreach',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {
    /**
     * New Account
     * method for OUTREACH_TRIGGER_ACCOUNT_CREATED
     */
    accountCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'OUTREACH_TRIGGER_ACCOUNT_CREATED',
        actionType: 'outreach',
        parameters: inputs,
      });
    },
    /**
     * Account Updated
     * method for OUTREACH_TRIGGER_ACCOUNT_UPDATED
     */
    accountUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'OUTREACH_TRIGGER_ACCOUNT_UPDATED',
        actionType: 'outreach',
        parameters: inputs,
      });
    },
    /**
     * New Opportunity
     * method for OUTREACH_TRIGGER_OPPORTUNITY_CREATED
     */
    opportunityCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'OUTREACH_TRIGGER_OPPORTUNITY_CREATED',
        actionType: 'outreach',
        parameters: inputs,
      });
    },
    /**
     * Opportunity Updated
     * method for OUTREACH_TRIGGER_OPPORTUNITY_UPDATED
     */
    opportunityUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'OUTREACH_TRIGGER_OPPORTUNITY_UPDATED',
        actionType: 'outreach',
        parameters: inputs,
      });
    },
    /**
     * New Prospect
     * method for OUTREACH_TRIGGER_PROSPECT_CREATED
     */
    prospectCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'OUTREACH_TRIGGER_PROSPECT_CREATED',
        actionType: 'outreach',
        parameters: inputs,
      });
    },
    /**
     * Prospect Updated
     * method for OUTREACH_TRIGGER_PROSPECT_UPDATED
     */
    prospectUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'OUTREACH_TRIGGER_PROSPECT_UPDATED',
        actionType: 'outreach',
        parameters: inputs,
      });
    },
  },
};
