// @useparagon/core/dist/steps/library/response/response.step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseStep = void 0;
const execution_1 = require("../../../execution");
const resolvers_1 = require("../../../resolvers");
const resolver_utils_1 = require("../../../resolvers/resolver.utils");
const step_1 = require("../../step");
const step_interface_1 = require("../../step.interface");
const response_interface_1 = require("./response.interface");
class ResponseStep extends step_1.Step {
    type = step_interface_1.StepType.RESPONSE;
    parameters;
    get output() {
        return (0, execution_1.createProxyInputReference)(this, () => this.id, execution_1.ProxyVariableSourceType.STEP_REFERENCE);
    }
    _parameters;
    constructor(params) {
        super(params);
        this._parameters = params;
        this.parameters = this.serializeParameters();
    }
    serializeParameters() {
        const initParams = this._parameters;
        if (initParams.responseType === response_interface_1.ResponseTypeEnum.FILE) {
            return {
                statusCode: initParams.statusCode,
                bodyData: [],
                responseType: response_interface_1.ResponseTypeEnum.FILE,
                bodyFile: (0, resolver_utils_1.resolveToSource)(initParams.body, resolvers_1.DataType.FILE, this.dependencies),
            };
        }
        return {
            statusCode: initParams.statusCode,
            responseType: response_interface_1.ResponseTypeEnum.JSON,
            bodyData: (0, resolver_utils_1.resolveParamsToSources)(initParams.body, undefined, undefined, this.dependencies),
        };
    }
}
exports.ResponseStep = ResponseStep;
