// @useparagon/integrations/salesforce/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for salesforce
 */
exports.inputs = [
  {
    id: 'owners',
    title: 'Account Owner',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'leadStatuses',
    title: 'Lead Status',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'opportunityStages',
    title: 'Opportunity Stage',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'customObjectMapping',
    title: 'Field Mapping',
    type: 'FIELD_MAPPER_DATA_SOURCE',
    fieldSource: {
      id: 'cachedFields',
      title: 'Field',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    recordSource: {
      id: 'recordTypes',
      title: 'Record Type',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
  {
    id: 'campaign',
    title: 'Campaign',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'recordTypeViewCombo',
    title: 'List View',
    type: 'COMBO_INPUT_DATA_SOURCE',
    mainInputSource: {
      id: 'recordTypesForListView',
      title: 'Record Type',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    dependentInputSource: {
      id: 'listView',
      title: 'List View',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
];

/**
 * triggers for salesforce
 */
exports.triggers = [
  {
    title: 'New Record',
    name: 'SALESFORCE_TRIGGER_RECORD_CREATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'Opportunity',
          'Account',
          'Contact',
          'Lead',
          'Task',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'recordsFilterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on new records that match these filters.',
        supportedOperators: [
          '$none',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringDoesNotEndWith',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$booleanTrue',
          '$booleanFalse',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Record Deleted',
    name: 'SALESFORCE_TRIGGER_RECORD_DELETED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'Opportunity',
          'Account',
          'Contact',
          'Lead',
          'Task',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'recordsFilterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on deleted records that match these filters.',
        supportedOperators: [
          '$none',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringDoesNotEndWith',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$booleanTrue',
          '$booleanFalse',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Record Updated',
    name: 'SALESFORCE_TRIGGER_RECORD_UPDATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'Opportunity',
          'Account',
          'Contact',
          'Lead',
          'Task',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object. Use Connect Portal Workflow Settings to allow users to select a Custom Object.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'recordsFilterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on updated records that match these filters.',
        supportedOperators: [
          '$none',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringDoesNotEndWith',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$booleanTrue',
          '$booleanFalse',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Campaign Member Added',
    name: 'SALESFORCE_TRIGGER_CAMPAIGN_MEMBER_ADDED',
    output: [],
    inputs: [
      {
        id: 'campaignId',
        title: 'Campaign',
        type: 'TEXTAREA',
        subtitle:
          'The campaign to watch for new Campaign Members. Use Connect Portal Workflow Settings to allow users to select a campaign.',
        placeholder: '{{settings.campaign}}',
        required: true,
      },
      {
        id: 'filterFormula',
        title: 'Filter records',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on records that match these filters.',
        placeholder: 'filter by column',
        supportedKeys: [
          'Id',
          'LeadId',
          'ContactId',
          'Status',
          'Type',
          'Name',
          'Email',
        ],
        supportedOperators: [
          '$none',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringDoesNotEndWith',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$booleanTrue',
          '$booleanFalse',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Campaign Opportunity Added',
    name: 'SALESFORCE_TRIGGER_CAMPAIGN_OPPORTUNITY_ADDED',
    output: [],
    inputs: [
      {
        id: 'campaignId',
        title: 'Campaign',
        type: 'TEXTAREA',
        subtitle:
          'The campaign to watch for new Campaign Members. Use Connect Portal Workflow Settings to allow users to select a campaign.',
        placeholder: '{{settings.campaign}}',
        required: true,
      },
      {
        id: 'filterFormula',
        title: 'Filter records',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on records that match these filters.',
        placeholder: 'filter by column',
        supportedKeys: [
          'Id',
          'Name',
          'ContactId',
          'StageName',
          'Probability',
          'Type',
        ],
        supportedOperators: [
          '$none',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringDoesNotEndWith',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$booleanTrue',
          '$booleanFalse',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
];

/**
 * intents for salesforce
 */
exports.intents = [
  {
    title: 'Create record',
    name: 'SALESFORCE_CREATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'field-Name',
                title: 'Name',
                type: 'TEXTAREA',
                subtitle: 'The Opportunity name. This field is required.',
                required: true,
              },
              {
                id: 'field-StageName',
                title: 'Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Opportunity Stage if not provided. Use Connect Portal Workflow Settings to allow users to select an Opportunity Stage.',
              },
              {
                id: 'field-CloseDate',
                title: 'Close Date',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to 30 days from the current date if not provided.',
                placeholder: 'YYYY-MM-DD',
              },
              {
                id: 'field-AccountId',
                title: 'Account ID',
                type: 'TEXTAREA',
                subtitle: 'The Account that the Opportunity belongs to.',
              },
              {
                id: 'field-Amount',
                title: 'Amount',
                type: 'TEXTAREA',
                subtitle: 'Estimated total sale amount.',
              },
              {
                id: 'field-Description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'A description of the Opportunity.',
              },
              {
                id: 'field-OwnerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Salesforce user assigned to work on this Opportunity.',
              },
              {
                id: 'field-NextStep',
                title: 'Next Step',
                type: 'TEXTAREA',
                subtitle: 'Description of next task in closing Opportunity.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Opportunity fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'field-Name',
                title: 'Account Name',
                type: 'TEXTAREA',
                subtitle: 'The Account name. This field is required.',
                required: true,
              },
              {
                id: 'field-OwnerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Salesforce user assigned to this Account. Use Connect Portal Workflow Settings to allow users to select an Owner.',
              },
              {
                id: 'field-Website',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Account fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'field-FirstName',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-LastName',
                title: 'Last Name',
                type: 'TEXTAREA',
                subtitle: 'This field is required.',
                required: true,
              },
              {
                id: 'field-accountId',
                title: 'Account ID',
                type: 'TEXTAREA',
                subtitle: 'The Account that the Contact belongs to.',
              },
              {
                id: 'field-Email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle:
                  'Title of the contact, such as CEO or Vice President.',
              },
              {
                id: 'field-Description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'A description of the Contact.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'field-FirstName',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-LastName',
                title: 'Last Name',
                type: 'TEXTAREA',
                subtitle: 'This field is required.',
                required: true,
              },
              {
                id: 'field-Company',
                title: 'Company',
                type: 'TEXTAREA',
                subtitle: 'This field is required.',
                required: true,
              },
              {
                id: 'field-Email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Website',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle:
                  'Title of the contact, such as CEO or Vice President.',
              },
              {
                id: 'field-Status',
                title: 'Status',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lead Status if not provided. Use Connect Portal Workflow Settings to allow users to select an Lead Status.',
              },
              {
                id: 'field-Description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'A description of the Lead.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Lead fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'field-whatId',
                title: 'Related to ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Account or Opportunity this Task is related to.',
              },
              {
                id: 'field-whoId',
                title: 'Name ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Contact or Lead this Task is related to.',
              },
              {
                id: 'field-subject',
                title: 'Subject',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'field-activityDate',
                title: 'Activity Date',
                type: 'TEXTAREA',
                placeholder: 'YYYY-MM-DD',
              },
              {
                id: 'field-description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'A description of the Task.',
              },
              {
                id: 'field-taskSubtype',
                title: 'Task Subtype',
                type: 'ENUM',
                required: true,
                defaultValue: 'task',
                values: ['task', 'email', 'listEmail', 'call'],
              },
              {
                id: 'field-Status',
                title: 'Status',
                type: 'ENUM',
                values: ['Not Started', 'In Progress', 'Completed'],
              },
              {
                id: 'field-ownerId',
                title: 'Assigned To ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Salesforce user assigned to this Task. Use Connect Portal Workflow Settings to allow users to select an Owner.',
                placeholder: '{{settings.TaskAssignee}}',
              },
              {
                id: 'field-callDurationInSeconds',
                title: 'Call Duration',
                type: 'TEXTAREA',
              },
              {
                id: 'field-isReminderSet',
                title: 'Reminder Set',
                type: 'BOOLEAN_INPUT',
              },
              {
                id: 'field-reminderDateTime',
                title: 'Reminder Date/Time',
                type: 'TEXTAREA',
                placeholder: '2019-09-15T02:06:58.147Z',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Task fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to create. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'additionalFields',
                title: 'Record Fields',
                type: 'CODE',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Field Mapping to specify which fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Update record',
    name: 'SALESFORCE_UPDATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the record to update.',
                required: true,
              },
              {
                id: 'field-Name',
                title: 'Name',
                type: 'TEXTAREA',
                subtitle: 'The Opportunity name.',
              },
              {
                id: 'field-StageName',
                title: 'Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Opportunity Stage if not provided. Use Connect Portal Workflow Settings to allow users to select an Opportunity Stage.',
              },
              {
                id: 'field-CloseDate',
                title: 'Close Date',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to 30 days from the current date if not provided.',
                placeholder: 'YYYY-MM-DD',
              },
              {
                id: 'field-AccountId',
                title: 'Account ID',
                type: 'TEXTAREA',
                subtitle: 'The Account that the Opportunity belongs to.',
              },
              {
                id: 'field-Amount',
                title: 'Amount',
                type: 'TEXTAREA',
                subtitle: 'Estimated total sale amount.',
              },
              {
                id: 'field-Description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'A description of the Opportunity.',
              },
              {
                id: 'field-OwnerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Salesforce user assigned to work on this Opportunity.',
              },
              {
                id: 'field-NextStep',
                title: 'Next Step',
                type: 'TEXTAREA',
                subtitle: 'Description of next task in closing Opportunity.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Opportunity fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the record to update.',
                required: true,
              },
              {
                id: 'field-Name',
                title: 'Account Name',
                type: 'TEXTAREA',
                subtitle: 'The Account name.',
              },
              {
                id: 'field-OwnerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Salesforce user assigned to this Account. Use Connect Portal Workflow Settings to allow users to select an Owner.',
              },
              {
                id: 'field-Website',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Account fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the record to update.',
                required: true,
              },
              {
                id: 'field-FirstName',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-LastName',
                title: 'Last Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-accountId',
                title: 'Account ID',
                type: 'TEXTAREA',
                subtitle: 'The Account that the Contact belongs to.',
              },
              {
                id: 'field-Email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle:
                  'Title of the contact, such as CEO or Vice President.',
              },
              {
                id: 'field-Description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'A description of the Contact.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the record to update.',
                required: true,
              },
              {
                id: 'field-FirstName',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-LastName',
                title: 'Last Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Company',
                title: 'Company',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Website',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'field-Title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle:
                  'Title of the contact, such as CEO or Vice President.',
              },
              {
                id: 'field-Status',
                title: 'Status',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lead Status if not provided. Use Connect Portal Workflow Settings to allow users to select an Lead Status.',
              },
              {
                id: 'field-Description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'A description of the Lead.',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Lead fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the record to update.',
                required: true,
              },
              {
                id: 'field-whatId',
                title: 'Related to ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Account or Opportunity this Task is related to.',
              },
              {
                id: 'field-whoId',
                title: 'Name ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Contact or Lead this Task is related to.',
              },
              {
                id: 'field-subject',
                title: 'Subject',
                type: 'TEXTAREA',
              },
              {
                id: 'field-activityDate',
                title: 'Activity Date',
                type: 'TEXTAREA',
                placeholder: 'YYYY-MM-DD',
              },
              {
                id: 'field-description',
                title: 'Description',
                type: 'TEXTAREA',
                subtitle: 'A description of the Task.',
              },
              {
                id: 'field-Status',
                title: 'Status',
                type: 'ENUM',
                values: ['Not Started', 'In Progress', 'Completed'],
              },
              {
                id: 'field-ownerId',
                title: 'Assigned To ID',
                type: 'TEXTAREA',
                subtitle:
                  'The Salesforce user assigned to this Task. Use Connect Portal Workflow Settings to allow users to select an Owner.',
                placeholder: '{{settings.TaskAssignee}}',
              },
              {
                id: 'field-callDurationInSeconds',
                title: 'Call Duration',
                type: 'TEXTAREA',
              },
              {
                id: 'field-isReminderSet',
                title: 'Reminder Set',
                type: 'BOOLEAN_INPUT',
              },
              {
                id: 'field-reminderDateTime',
                title: 'Reminder Date/Time',
                type: 'TEXTAREA',
                placeholder: '2019-09-15T02:06:58.147Z',
              },
              {
                id: 'additionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Task fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to create. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the record to update.',
                required: true,
              },
              {
                id: 'additionalFields',
                title: 'Record Fields',
                type: 'CODE',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Field Mapping to specify which fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get record by ID',
    name: 'SALESFORCE_GET_RECORD_BY_ID',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to create. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Search records',
    name: 'SALESFORCE_SEARCH_RECORDS',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'Name',
                  'StageName',
                  'AccountId',
                  'Amount',
                  'OwnerId',
                  'NextStep',
                ],
                supportedOperators: [
                  '$none',
                  '$stringContains',
                  '$stringDoesNotContain',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringIsIn',
                  '$stringIsNotIn',
                  '$stringStartsWith',
                  '$stringDoesNotStartWith',
                  '$stringEndsWith',
                  '$stringDoesNotEndWith',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$booleanTrue',
                  '$booleanFalse',
                  '$exists',
                  '$doesNotExist',
                ],
              },
              {
                id: 'sortBy',
                title: 'Sort By',
                type: 'TEXTAREA',
                placeholder: 'CreatedDate',
              },
              {
                id: 'sortDirection',
                title: 'Sort Direction',
                type: 'ENUM',
                values: ['ASC', 'DESC'],
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'Name',
                  'OwnerId',
                  'Website',
                  'Phone',
                  'Description',
                ],
                supportedOperators: [
                  '$none',
                  '$stringContains',
                  '$stringDoesNotContain',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringIsIn',
                  '$stringIsNotIn',
                  '$stringStartsWith',
                  '$stringDoesNotStartWith',
                  '$stringEndsWith',
                  '$stringDoesNotEndWith',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$booleanTrue',
                  '$booleanFalse',
                  '$exists',
                  '$doesNotExist',
                ],
              },
              {
                id: 'sortBy',
                title: 'Sort By',
                type: 'TEXTAREA',
                placeholder: 'CreatedDate',
              },
              {
                id: 'sortDirection',
                title: 'Sort Direction',
                type: 'ENUM',
                values: ['ASC', 'DESC'],
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'FirstName',
                  'LastName',
                  'accountId',
                  'Email',
                  'Title',
                  'Description',
                ],
                supportedOperators: [
                  '$none',
                  '$stringContains',
                  '$stringDoesNotContain',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringIsIn',
                  '$stringIsNotIn',
                  '$stringStartsWith',
                  '$stringDoesNotStartWith',
                  '$stringEndsWith',
                  '$stringDoesNotEndWith',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$booleanTrue',
                  '$booleanFalse',
                  '$exists',
                  '$doesNotExist',
                ],
              },
              {
                id: 'sortBy',
                title: 'Sort By',
                type: 'TEXTAREA',
                placeholder: 'CreatedDate',
              },
              {
                id: 'sortDirection',
                title: 'Sort Direction',
                type: 'ENUM',
                values: ['ASC', 'DESC'],
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'FirstName',
                  'LastName',
                  'Company',
                  'Email',
                  'Phone',
                  'Website',
                  'Title',
                  'Status',
                  'Description',
                ],
                supportedOperators: [
                  '$none',
                  '$stringContains',
                  '$stringDoesNotContain',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringIsIn',
                  '$stringIsNotIn',
                  '$stringStartsWith',
                  '$stringDoesNotStartWith',
                  '$stringEndsWith',
                  '$stringDoesNotEndWith',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$booleanTrue',
                  '$booleanFalse',
                  '$exists',
                  '$doesNotExist',
                ],
              },
              {
                id: 'sortBy',
                title: 'Sort By',
                type: 'TEXTAREA',
                placeholder: 'CreatedDate',
              },
              {
                id: 'sortDirection',
                title: 'Sort Direction',
                type: 'ENUM',
                values: ['ASC', 'DESC'],
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'subject',
                  'taskSubtype',
                  'Status',
                  'ownerId',
                  'callDurationInSeconds',
                  'reminderDateTime',
                ],
                supportedOperators: [
                  '$none',
                  '$stringContains',
                  '$stringDoesNotContain',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringIsIn',
                  '$stringIsNotIn',
                  '$stringStartsWith',
                  '$stringDoesNotStartWith',
                  '$stringEndsWith',
                  '$stringDoesNotEndWith',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$booleanTrue',
                  '$booleanFalse',
                  '$exists',
                  '$doesNotExist',
                ],
              },
              {
                id: 'sortBy',
                title: 'Sort By',
                type: 'TEXTAREA',
                placeholder: 'CreatedDate',
              },
              {
                id: 'sortDirection',
                title: 'Sort Direction',
                type: 'ENUM',
                values: ['ASC', 'DESC'],
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to create. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedOperators: [
                  '$none',
                  '$stringContains',
                  '$stringDoesNotContain',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$stringIsIn',
                  '$stringIsNotIn',
                  '$stringStartsWith',
                  '$stringDoesNotStartWith',
                  '$stringEndsWith',
                  '$stringDoesNotEndWith',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$booleanTrue',
                  '$booleanFalse',
                  '$exists',
                  '$doesNotExist',
                ],
              },
              {
                id: 'sortBy',
                title: 'Sort By',
                type: 'TEXTAREA',
                placeholder: 'CreatedDate',
              },
              {
                id: 'sortDirection',
                title: 'Sort Direction',
                type: 'ENUM',
                values: ['ASC', 'DESC'],
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Search records by SOQL query',
    name: 'SALESFORCE_WRITE_SOQL_QUERY',
    output: [],
    inputs: [
      {
        id: 'query',
        title: 'SOQL Query',
        type: 'CODE',
        placeholder: "SELECT Id, Name\nFROM Account\nWHERE Name = 'Ethan'",
        required: true,
      },
    ],
  },
  {
    title: 'Deploy custom field',
    name: 'SALESFORCE_CREATE_CUSTOM_FIELD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Opportunity',
            dependentInputs: [
              {
                id: 'label',
                title: 'Field Label',
                type: 'TEXTAREA',
                subtitle:
                  'Enter a label to be used on displays, page layouts, reports, and list views. Also determines the Field Name used for internal reference. For example, if the Field Label is “Email”, the Field Name will be “Email__c”.',
                required: true,
              },
              {
                id: 'type',
                title: 'Field Type',
                type: 'ENUM',
                required: true,
                defaultValue: 'Text',
                values: [
                  {
                    value: 'Checkbox',
                    dependentInputs: [
                      {
                        id: 'defaultCheckboxValue',
                        title: 'Default Value',
                        type: 'BOOLEAN_INPUT',
                        defaultValue: false,
                      },
                    ],
                  },
                  {
                    value: 'Currency',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Date',
                  'Email',
                  {
                    value: 'Number',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Percent',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Phone',
                  {
                    value: 'Picklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'MultiselectPicklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Text',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                    ],
                  },
                  'TextArea',
                  {
                    value: 'LongTextArea',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Html',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'The maximum length of the text field. Minimum length 256',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        subtitle: 'undefined Minimum 10',
                        required: true,
                      },
                    ],
                  },
                  'Time',
                  'Url',
                ],
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'helperText',
                title: 'Helper Text',
                type: 'TEXTAREA',
                subtitle:
                  'This text displays on detail and edit pages when users hover over the Info icon next to this field.',
              },
              {
                id: 'defaultFieldValue',
                title: 'Default Value',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: 'Account',
            dependentInputs: [
              {
                id: 'label',
                title: 'Field Label',
                type: 'TEXTAREA',
                subtitle:
                  'Enter a label to be used on displays, page layouts, reports, and list views. Also determines the Field Name used for internal reference. For example, if the Field Label is “Email”, the Field Name will be “Email__c”.',
                required: true,
              },
              {
                id: 'type',
                title: 'Field Type',
                type: 'ENUM',
                required: true,
                defaultValue: 'Text',
                values: [
                  {
                    value: 'Checkbox',
                    dependentInputs: [
                      {
                        id: 'defaultCheckboxValue',
                        title: 'Default Value',
                        type: 'BOOLEAN_INPUT',
                        defaultValue: false,
                      },
                    ],
                  },
                  {
                    value: 'Currency',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Date',
                  'Email',
                  {
                    value: 'Number',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Percent',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Phone',
                  {
                    value: 'Picklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'MultiselectPicklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Text',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                    ],
                  },
                  'TextArea',
                  {
                    value: 'LongTextArea',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Html',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'The maximum length of the text field. Minimum length 256',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        subtitle: 'undefined Minimum 10',
                        required: true,
                      },
                    ],
                  },
                  'Time',
                  'Url',
                ],
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'helperText',
                title: 'Helper Text',
                type: 'TEXTAREA',
                subtitle:
                  'This text displays on detail and edit pages when users hover over the Info icon next to this field.',
              },
              {
                id: 'defaultFieldValue',
                title: 'Default Value',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: 'Contact',
            dependentInputs: [
              {
                id: 'label',
                title: 'Field Label',
                type: 'TEXTAREA',
                subtitle:
                  'Enter a label to be used on displays, page layouts, reports, and list views. Also determines the Field Name used for internal reference. For example, if the Field Label is “Email”, the Field Name will be “Email__c”.',
                required: true,
              },
              {
                id: 'type',
                title: 'Field Type',
                type: 'ENUM',
                required: true,
                defaultValue: 'Text',
                values: [
                  {
                    value: 'Checkbox',
                    dependentInputs: [
                      {
                        id: 'defaultCheckboxValue',
                        title: 'Default Value',
                        type: 'BOOLEAN_INPUT',
                        defaultValue: false,
                      },
                    ],
                  },
                  {
                    value: 'Currency',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Date',
                  'Email',
                  {
                    value: 'Number',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Percent',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Phone',
                  {
                    value: 'Picklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'MultiselectPicklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Text',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                    ],
                  },
                  'TextArea',
                  {
                    value: 'LongTextArea',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Html',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'The maximum length of the text field. Minimum length 256',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        subtitle: 'undefined Minimum 10',
                        required: true,
                      },
                    ],
                  },
                  'Time',
                  'Url',
                ],
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'helperText',
                title: 'Helper Text',
                type: 'TEXTAREA',
                subtitle:
                  'This text displays on detail and edit pages when users hover over the Info icon next to this field.',
              },
              {
                id: 'defaultFieldValue',
                title: 'Default Value',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: 'Lead',
            dependentInputs: [
              {
                id: 'label',
                title: 'Field Label',
                type: 'TEXTAREA',
                subtitle:
                  'Enter a label to be used on displays, page layouts, reports, and list views. Also determines the Field Name used for internal reference. For example, if the Field Label is “Email”, the Field Name will be “Email__c”.',
                required: true,
              },
              {
                id: 'type',
                title: 'Field Type',
                type: 'ENUM',
                required: true,
                defaultValue: 'Text',
                values: [
                  {
                    value: 'Checkbox',
                    dependentInputs: [
                      {
                        id: 'defaultCheckboxValue',
                        title: 'Default Value',
                        type: 'BOOLEAN_INPUT',
                        defaultValue: false,
                      },
                    ],
                  },
                  {
                    value: 'Currency',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Date',
                  'Email',
                  {
                    value: 'Number',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Percent',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Phone',
                  {
                    value: 'Picklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'MultiselectPicklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Text',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                    ],
                  },
                  'TextArea',
                  {
                    value: 'LongTextArea',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Html',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'The maximum length of the text field. Minimum length 256',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        subtitle: 'undefined Minimum 10',
                        required: true,
                      },
                    ],
                  },
                  'Time',
                  'Url',
                ],
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'helperText',
                title: 'Helper Text',
                type: 'TEXTAREA',
                subtitle:
                  'This text displays on detail and edit pages when users hover over the Info icon next to this field.',
              },
              {
                id: 'defaultFieldValue',
                title: 'Default Value',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: 'Task',
            dependentInputs: [
              {
                id: 'label',
                title: 'Field Label',
                type: 'TEXTAREA',
                subtitle:
                  'Enter a label to be used on displays, page layouts, reports, and list views. Also determines the Field Name used for internal reference. For example, if the Field Label is “Email”, the Field Name will be “Email__c”.',
                required: true,
              },
              {
                id: 'type',
                title: 'Field Type',
                type: 'ENUM',
                required: true,
                defaultValue: 'Text',
                values: [
                  {
                    value: 'Checkbox',
                    dependentInputs: [
                      {
                        id: 'defaultCheckboxValue',
                        title: 'Default Value',
                        type: 'BOOLEAN_INPUT',
                        defaultValue: false,
                      },
                    ],
                  },
                  {
                    value: 'Currency',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Date',
                  'Email',
                  {
                    value: 'Number',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Percent',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Phone',
                  {
                    value: 'Picklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'MultiselectPicklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Text',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                    ],
                  },
                  'TextArea',
                  'Time',
                  'Url',
                ],
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'helperText',
                title: 'Helper Text',
                type: 'TEXTAREA',
                subtitle:
                  'This text displays on detail and edit pages when users hover over the Info icon next to this field.',
              },
              {
                id: 'defaultFieldValue',
                title: 'Default Value',
                type: 'TEXTAREA',
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to create. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'label',
                title: 'Field Label',
                type: 'TEXTAREA',
                subtitle:
                  'Enter a label to be used on displays, page layouts, reports, and list views. Also determines the Field Name used for internal reference. For example, if the Field Label is “Email”, the Field Name will be “Email__c”.',
                required: true,
              },
              {
                id: 'type',
                title: 'Field Type',
                type: 'ENUM',
                required: true,
                defaultValue: 'Text',
                values: [
                  {
                    value: 'Checkbox',
                    dependentInputs: [
                      {
                        id: 'defaultCheckboxValue',
                        title: 'Default Value',
                        type: 'BOOLEAN_INPUT',
                        defaultValue: false,
                      },
                    ],
                  },
                  {
                    value: 'Currency',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Date',
                  'Email',
                  {
                    value: 'Number',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Percent',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the left of the decimal point',
                        required: true,
                      },
                      {
                        id: 'decimalPlace',
                        title: 'Decimal Places',
                        type: 'TEXTAREA',
                        subtitle:
                          'Number of digits to the right of the decimal point',
                        required: true,
                      },
                    ],
                  },
                  'Phone',
                  {
                    value: 'Picklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'MultiselectPicklist',
                    dependentInputs: [
                      {
                        id: 'pickListValues',
                        title: 'Values',
                        type: 'TEXTAREA',
                        subtitle:
                          'Enter values, with each value separated by a new line',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Text',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                    ],
                  },
                  'TextArea',
                  {
                    value: 'LongTextArea',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle: 'The maximum length of the text field.',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        required: true,
                      },
                    ],
                  },
                  {
                    value: 'Html',
                    dependentInputs: [
                      {
                        id: 'length',
                        title: 'Length',
                        type: 'TEXTAREA',
                        subtitle:
                          'The maximum length of the text field. Minimum length 256',
                        required: true,
                      },
                      {
                        id: 'visibleLines',
                        title: 'Visible Lines',
                        type: 'TEXTAREA',
                        subtitle: 'undefined Minimum 10',
                        required: true,
                      },
                    ],
                  },
                  'Time',
                  'Url',
                ],
              },
              {
                id: 'description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'helperText',
                title: 'Helper Text',
                type: 'TEXTAREA',
                subtitle:
                  'This text displays on detail and edit pages when users hover over the Info icon next to this field.',
              },
              {
                id: 'defaultFieldValue',
                title: 'Default Value',
                type: 'TEXTAREA',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Deploy custom object',
    name: 'SALESFORCE_CREATE_CUSTOM_OBJECT',
    output: [],
    inputs: [
      {
        id: 'label',
        title: 'Label',
        type: 'TEXTAREA',
        subtitle:
          'Object Label to be used in tabs, page layouts, and reports. Also determines the Object Name used for internal reference. For example, if the Label is “Account”, the Object Name will be “Account__c”.',
        placeholder: 'Account',
        required: true,
      },
      {
        id: 'pluralLabel',
        title: 'Plural Label',
        type: 'TEXTAREA',
        placeholder: 'Accounts',
        required: true,
      },
      {
        id: 'description',
        title: 'Description',
        type: 'TEXTAREA',
        subtitle: 'A description of the Custom Object.',
      },
      {
        id: 'recordName',
        title: 'Record Name',
        type: 'TEXTAREA',
        subtitle:
          'The Record Name appears in page layouts, key lists, related lists, lookups, and search results. For example, the Record Name for Account is "Account Name" and for Case it is "Case Number".',
        placeholder: 'Account Name',
        required: true,
      },
    ],
  },
  {
    title: 'Get records by List View',
    name: 'SALESFORCE_GET_CONTACTS_BY_VIEW_ID',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'EDITABLE_ENUM',
        subtitle:
          'Use Connect Portal User Settings to allow users to select a Record Type.',
        required: true,
        values: ['Opportunity', 'Account', 'Contact', 'Lead', 'Task'],
      },
      {
        id: 'listViewId',
        title: 'List View ID',
        type: 'TEXTAREA',
        required: true,
      },
    ],
  },
];

/**
 * integration class for salesforce
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'salesforce',

  /**
   * action step methods
   */
  actions: {
    /**
     * Create record
     * method for SALESFORCE_CREATE_RECORD
     */
    createRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'SALESFORCE_CREATE_RECORD',
        actionType: 'salesforce',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update record
     * method for SALESFORCE_UPDATE_RECORD
     */
    updateRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'SALESFORCE_UPDATE_RECORD',
        actionType: 'salesforce',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get record by ID
     * method for SALESFORCE_GET_RECORD_BY_ID
     */
    getRecordById: (inputs, options) => {
      return new ActionStep({
        intent: 'SALESFORCE_GET_RECORD_BY_ID',
        actionType: 'salesforce',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Search records
     * method for SALESFORCE_SEARCH_RECORDS
     */
    searchRecords: (inputs, options) => {
      return new ActionStep({
        intent: 'SALESFORCE_SEARCH_RECORDS',
        actionType: 'salesforce',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Search records by SOQL query
     * method for SALESFORCE_WRITE_SOQL_QUERY
     */
    writeSoqlQuery: (inputs, options) => {
      return new ActionStep({
        intent: 'SALESFORCE_WRITE_SOQL_QUERY',
        actionType: 'salesforce',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Deploy custom field
     * method for SALESFORCE_CREATE_CUSTOM_FIELD
     */
    createCustomField: (inputs, options) => {
      return new ActionStep({
        intent: 'SALESFORCE_CREATE_CUSTOM_FIELD',
        actionType: 'salesforce',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Deploy custom object
     * method for SALESFORCE_CREATE_CUSTOM_OBJECT
     */
    createCustomObject: (inputs, options) => {
      return new ActionStep({
        intent: 'SALESFORCE_CREATE_CUSTOM_OBJECT',
        actionType: 'salesforce',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get records by List View
     * method for SALESFORCE_GET_CONTACTS_BY_VIEW_ID
     */
    getContactsByViewId: (inputs, options) => {
      return new ActionStep({
        intent: 'SALESFORCE_GET_CONTACTS_BY_VIEW_ID',
        actionType: 'salesforce',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {
    /**
     * New Record
     * method for SALESFORCE_TRIGGER_RECORD_CREATED
     */
    recordCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'SALESFORCE_TRIGGER_RECORD_CREATED',
        actionType: 'salesforce',
        parameters: inputs,
      });
    },
    /**
     * Record Deleted
     * method for SALESFORCE_TRIGGER_RECORD_DELETED
     */
    recordDeleted: (inputs) => {
      return new ActionTriggerStep({
        intent: 'SALESFORCE_TRIGGER_RECORD_DELETED',
        actionType: 'salesforce',
        parameters: inputs,
      });
    },
    /**
     * Record Updated
     * method for SALESFORCE_TRIGGER_RECORD_UPDATED
     */
    recordUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'SALESFORCE_TRIGGER_RECORD_UPDATED',
        actionType: 'salesforce',
        parameters: inputs,
      });
    },
    /**
     * Campaign Member Added
     * method for SALESFORCE_TRIGGER_CAMPAIGN_MEMBER_ADDED
     */
    campaignMemberAdded: (inputs) => {
      return new ActionTriggerStep({
        intent: 'SALESFORCE_TRIGGER_CAMPAIGN_MEMBER_ADDED',
        actionType: 'salesforce',
        parameters: inputs,
      });
    },
    /**
     * Campaign Opportunity Added
     * method for SALESFORCE_TRIGGER_CAMPAIGN_OPPORTUNITY_ADDED
     */
    campaignOpportunityAdded: (inputs) => {
      return new ActionTriggerStep({
        intent: 'SALESFORCE_TRIGGER_CAMPAIGN_OPPORTUNITY_ADDED',
        actionType: 'salesforce',
        parameters: inputs,
      });
    },
  },
};
