// @useparagon/core/dist/steps/library/action/action.step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionStep = void 0;
const execution_1 = require("../../../execution");
const resolver_utils_1 = require("../../../resolvers/resolver.utils");
const resolvers_interface_1 = require("../../../resolvers/resolvers.interface");
const step_1 = require("../../step");
const step_interface_1 = require("../../step.interface");
class ActionStep extends step_1.Step {
    type = step_interface_1.StepType.ACTION;
    parameters;
    _parameters;
    get output() {
        return (0, execution_1.createProxyInputReference)(this, () => this.id, execution_1.ProxyVariableSourceType.STEP_REFERENCE);
    }
    constructor(params) {
        super(params);
        this._parameters = params;
        this.parameters = this.serializeParameters();
    }
    serializeParameters() {
        const initParams = this._parameters;
        const serializedActionParameters = (0, resolver_utils_1.resolveParamsToSources)(initParams.parameters, resolvers_interface_1.DataType.ANY, undefined, this.dependencies);
        const output = {
            intent: initParams.intent,
            actionType: initParams.actionType,
            credentials: [],
            actionParameters: serializedActionParameters,
            ignoreFailure: Boolean(initParams.continueWorkflowOnError),
            retryOnFailure: Boolean(initParams.autoRetry),
        };
        return output;
    }
}
exports.ActionStep = ActionStep;
