// @useparagon/core/dist/resource/resource.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppLevelAPIResource = exports.UserLevelAPIResource = exports.OAUTH_CLIENT_CREDENTIAL_REQUEST_TYPE = exports.ResourceAuthorizationType = exports.ResourceConnectionScope = exports.ResourceCategory = void 0;
var ResourceCategory;
(function (ResourceCategory) {
    ResourceCategory["API"] = "api";
})(ResourceCategory || (exports.ResourceCategory = ResourceCategory = {}));
var ResourceConnectionScope;
(function (ResourceConnectionScope) {
    ResourceConnectionScope["APP"] = "APP";
    ResourceConnectionScope["USER"] = "USER";
})(ResourceConnectionScope || (exports.ResourceConnectionScope = ResourceConnectionScope = {}));
var ResourceAuthorizationType;
(function (ResourceAuthorizationType) {
    ResourceAuthorizationType["NONE"] = "none";
    ResourceAuthorizationType["OAUTH"] = "oauth";
    ResourceAuthorizationType["OAUTH_CLIENT_CREDENTIAL"] = "oauth_client_credential";
    ResourceAuthorizationType["API_KEY"] = "api_key";
})(ResourceAuthorizationType || (exports.ResourceAuthorizationType = ResourceAuthorizationType = {}));
var OAUTH_CLIENT_CREDENTIAL_REQUEST_TYPE;
(function (OAUTH_CLIENT_CREDENTIAL_REQUEST_TYPE) {
    OAUTH_CLIENT_CREDENTIAL_REQUEST_TYPE["BodyParam"] = "body-param";
    OAUTH_CLIENT_CREDENTIAL_REQUEST_TYPE["BasicAuthHeader"] = "basic-auth-header";
})(OAUTH_CLIENT_CREDENTIAL_REQUEST_TYPE || (exports.OAUTH_CLIENT_CREDENTIAL_REQUEST_TYPE = OAUTH_CLIENT_CREDENTIAL_REQUEST_TYPE = {}));
class UserLevelAPIResource {
    connectionScope = ResourceConnectionScope.USER;
    category = ResourceCategory.API;
}
exports.UserLevelAPIResource = UserLevelAPIResource;
class AppLevelAPIResource {
    connectionScope = ResourceConnectionScope.APP;
    category = ResourceCategory.API;
}
exports.AppLevelAPIResource = AppLevelAPIResource;
