// @useparagon/core/dist/user/user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectUser = void 0;
const context_interface_1 = require("../execution/context.interface");
const context_utils_1 = require("../execution/context.utils");
class ConnectUser {
    get meta() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.PERSONA_META_REFERENCE, ['meta']);
    }
    get providerId() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CONNECT_USER_PROVIDER_ID);
    }
    get userId() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.PERSONA_ENDUSERID_REFERENCE);
    }
    get providerData() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CONNECT_USER_PROVIDER_DATA);
    }
    get credentialId() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CONNECT_USER_CREDENTIAL_ID);
    }
    get configurationId() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CONNECT_USER_CONFIGURATION_ID);
    }
}
exports.ConnectUser = ConnectUser;
