// @useparagon/core/dist/triggers/customTrigger/customTrigger.context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomTriggerContext = void 0;
const context_interface_1 = require("../../execution/context.interface");
const context_utils_1 = require("../../execution/context.utils");
class CustomTriggerContext {
    get webhookURL() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CUSTOM_WEBHOOK_TRIGGER_URL);
    }
    get setupResponse() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CUSTOM_WEBHOOK_TRIGGER_SETUP_RESPONSE);
    }
    get webhookRequest() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CUSTOM_WEBHOOK_TRIGGER_WEBHOOK_RESPONSE, ['body']);
    }
    get credential() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CREDENTIAL_VALUES);
    }
    get providerData() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CONNECT_USER_PROVIDER_DATA);
    }
    get profileData() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.PROFILE_CONFIG_KEY);
    }
    get persistData() {
        return (0, context_utils_1.createProxyInputReference)({}, () => '', context_interface_1.ProxyVariableSourceType.CUSTOM_WEBHOOK_TRIGGER_HANDSHAKE_RESPONSE);
    }
    getEnvironmentSecret(key) {
        return (0, context_utils_1.createProxyInputReference)(key, () => key, context_interface_1.ProxyVariableSourceType.SECRET_REFERENCE);
    }
    getInput(input) {
        return (0, context_utils_1.createProxyInputReference)(input, () => input.id, context_interface_1.ProxyVariableSourceType.WORKFLOW_INPUT_REFERENCE);
    }
}
exports.CustomTriggerContext = CustomTriggerContext;
