// @useparagon/core/dist/triggers/customTrigger/customTrigger.step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomTriggerStep = void 0;
const execution_1 = require("../../execution");
const step_interface_1 = require("../../steps/step.interface");
const trigger_1 = require("../trigger");
const customTrigger_interface_1 = require("./customTrigger.interface");
class CustomTriggerStep extends trigger_1.TriggerStep {
    get output() {
        return (0, execution_1.createProxyInputReference)(this, () => this.id, execution_1.ProxyVariableSourceType.STEP_REFERENCE);
    }
    _parameters;
    parameters;
    type = step_interface_1.StepType.ACTION_TRIGGER;
    constructor(CustomTriggerClass) {
        super({});
        this._parameters = new CustomTriggerClass();
        this.parameters = this.serializeParameters();
    }
    serializeParameters() {
        const customTriggerId = this._parameters.id;
        const customTrigger = (this.dependencies.customTriggers ?? []).find((trigger) => trigger.id === customTriggerId);
        if (this.dependencies.customTriggers && !customTrigger) {
            throw new Error(`Trigger with id "${customTriggerId}" not defined in this integration.`);
        }
        return {
            ...this.parameters,
            actionType: this.dependencies.integrationType,
            intent: customTrigger_interface_1.CUSTOM_TRIGGER_INTENT,
            customWebhookTriggerId: customTriggerId,
            resourceId: customTrigger?.resourceId,
            actionParameters: [],
            credentials: [],
        };
    }
}
exports.CustomTriggerStep = CustomTriggerStep;
