// @useparagon/core/dist/triggers/customTrigger/customTrigger.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomTriggerPayloadValidationScheme = exports.SetupResponseInitType = exports.SetupResponseType = exports.CUSTOM_TRIGGER_INTENT = exports.TriggerSubscription = void 0;
var TriggerSubscription;
(function (TriggerSubscription) {
    TriggerSubscription["PROGRAMMATIC"] = "PROGRAMMATIC";
    TriggerSubscription["MANUAL"] = "MANUAL";
})(TriggerSubscription || (exports.TriggerSubscription = TriggerSubscription = {}));
exports.CUSTOM_TRIGGER_INTENT = 'CUSTOM_WEBHOOK_TRIGGER_INTENT';
var SetupResponseType;
(function (SetupResponseType) {
    SetupResponseType["DEFAULT"] = "DEFAULT";
    SetupResponseType["ECHO_REQUEST_BODY"] = "ECHO_REQUEST_BODY";
    SetupResponseType["CUSTOM_CODE"] = "CUSTOM_CODE";
})(SetupResponseType || (exports.SetupResponseType = SetupResponseType = {}));
var SetupResponseInitType;
(function (SetupResponseInitType) {
    SetupResponseInitType["RESPOND_WITH_200"] = "RESPOND_WITH_200";
    SetupResponseInitType["ECHO_REQUEST_BODY"] = "ECHO_REQUEST_BODY";
    SetupResponseInitType["CUSTOM_CODE"] = "CUSTOM_CODE";
})(SetupResponseInitType || (exports.SetupResponseInitType = SetupResponseInitType = {}));
var CustomTriggerPayloadValidationScheme;
(function (CustomTriggerPayloadValidationScheme) {
    CustomTriggerPayloadValidationScheme["NONE"] = "NONE";
    CustomTriggerPayloadValidationScheme["CUSTOM"] = "CUSTOM";
    CustomTriggerPayloadValidationScheme["SHA_256_HEX_PAYLOAD_BODY"] = "SHA_256_HEX_PAYLOAD_BODY";
    CustomTriggerPayloadValidationScheme["SHA_256_BASE64_PAYLOAD_BODY"] = "SHA_256_BASE64_PAYLOAD_BODY";
    CustomTriggerPayloadValidationScheme["SHA_1_HEX_PAYLOAD_BODY"] = "SHA_1_HEX_PAYLOAD_BODY";
    CustomTriggerPayloadValidationScheme["SHA_1_BASE64_PAYLOAD_BODY"] = "SHA_1_BASE64_PAYLOAD_BODY";
})(CustomTriggerPayloadValidationScheme || (exports.CustomTriggerPayloadValidationScheme = CustomTriggerPayloadValidationScheme = {}));
