// @useparagon/core/dist/resolvers/resolvers.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operator = exports.FanInStrategy = exports.DataType = void 0;
var DataType;
(function (DataType) {
    DataType["STRING"] = "STRING";
    DataType["NUMBER"] = "NUMBER";
    DataType["DATE"] = "DATE";
    DataType["BOOLEAN"] = "BOOLEAN";
    DataType["EMAIL"] = "EMAIL";
    DataType["OBJECT"] = "OBJECT";
    DataType["ARRAY"] = "ARRAY";
    DataType["ANY"] = "ANY";
    DataType["FILE"] = "FILE";
    DataType["NON_DECIMAL"] = "NON_DECIMAL";
})(DataType || (exports.DataType = DataType = {}));
var FanInStrategy;
(function (FanInStrategy) {
    FanInStrategy["SINGLE"] = "SINGLE";
    FanInStrategy["SINGLE_BY_FANOUT"] = "SINGLE_BY_FANOUT";
    FanInStrategy["MULTI"] = "MULTI";
    FanInStrategy["MULTI_BY_FANOUT"] = "MULTI_BY_FANOUT";
})(FanInStrategy || (exports.FanInStrategy = FanInStrategy = {}));
var Operator;
(function (Operator) {
    Operator["None"] = "$none";
    Operator["StringContains"] = "$stringContains";
    Operator["StringDoesNotContain"] = "$stringDoesNotContain";
    Operator["StringExactlyMatches"] = "$stringExactlyMatches";
    Operator["StringDoesNotExactlyMatch"] = "$stringDoesNotExactlyMatch";
    Operator["StringIsIn"] = "$stringIsIn";
    Operator["StringIsNotIn"] = "$stringIsNotIn";
    Operator["StringStartsWith"] = "$stringStartsWith";
    Operator["StringDoesNotStartWith"] = "$stringDoesNotStartWith";
    Operator["StringEndsWith"] = "$stringEndsWith";
    Operator["StringDoesNotEndWith"] = "$stringDoesNotEndWith";
    Operator["NumberGreaterThan"] = "$numberGreaterThan";
    Operator["NumberLessThan"] = "$numberLessThan";
    Operator["NumberEquals"] = "$numberEquals";
    Operator["NumberDoesNotEqual"] = "$numberDoesNotEqual";
    Operator["NumberLessThanOrEqualTo"] = "$numberLessThanOrEqualTo";
    Operator["NumberGreaterThanOrEqualTo"] = "$numberGreaterThanOrEqualTo";
    Operator["DateTimeAfter"] = "$dateTimeAfter";
    Operator["DateTimeBefore"] = "$dateTimeBefore";
    Operator["DateTimeEquals"] = "$dateTimeEquals";
    Operator["BooleanTrue"] = "$booleanTrue";
    Operator["BooleanFalse"] = "$booleanFalse";
    Operator["IsNotNull"] = "$exists";
    Operator["IsNull"] = "$doesNotExist";
    Operator["Exists"] = "$isNotUndefinedOrNull";
    Operator["DoesNotExist"] = "$isUndefinedOrNull";
    Operator["ArrayIsIn"] = "$arrayIsIn";
    Operator["ArrayIsNotIn"] = "$arrayIsNotIn";
    Operator["ArrayIsEmpty"] = "$arrayIsEmpty";
    Operator["ArrayIsNotEmpty"] = "$arrayIsNotEmpty";
    Operator["StringGreaterThan"] = "$stringGreaterThan";
    Operator["StringLessThan"] = "$stringLessThan";
})(Operator || (exports.Operator = Operator = {}));
