// @useparagon/core/dist/triggers/actionTrigger/actionTrigger.step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionTriggerStep = void 0;
const execution_1 = require("../../execution");
const resolver_utils_1 = require("../../resolvers/resolver.utils");
const resolvers_interface_1 = require("../../resolvers/resolvers.interface");
const step_interface_1 = require("../../steps/step.interface");
const trigger_1 = require("../trigger");
class ActionTriggerStep extends trigger_1.TriggerStep {
    type = step_interface_1.StepType.ACTION_TRIGGER;
    parameters;
    get output() {
        return (0, execution_1.createProxyInputReference)(this, () => this.id, execution_1.ProxyVariableSourceType.STEP_REFERENCE);
    }
    _parameters;
    constructor(params) {
        super(params);
        this._parameters = params;
        this.parameters = this.serializeParameters();
    }
    serializeParameters() {
        const { objectMapping, ...actionParameters } = this._parameters.parameters;
        const serializedActionParameters = (0, resolver_utils_1.resolveParamsToSources)(actionParameters, resolvers_interface_1.DataType.ANY, undefined, this.dependencies);
        if (objectMapping) {
            serializedActionParameters.push(...(0, resolver_utils_1.resolveParamsToSources)({ objectMapping }, resolvers_interface_1.DataType.STRING, true, this.dependencies));
        }
        const output = {
            ...this._parameters,
            credentials: [],
            actionParameters: serializedActionParameters,
        };
        delete output['parameters'];
        return output;
    }
}
exports.ActionTriggerStep = ActionTriggerStep;
