// @useparagon/integrations/hubspot/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for hubspot
 */
exports.inputs = [
  {
    id: 'dealStageByPipelineId',
    title: 'Deal Stage',
    type: 'COMBO_INPUT_DATA_SOURCE',
    mainInputSource: {
      id: 'pipelines',
      title: 'Pipelines',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    dependentInputSource: {
      id: 'DealStagesByPipeline',
      title: 'Deal Stages By Pipeline',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
  {
    id: 'leadStatuses',
    title: 'Company Lead Status',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'contactLeadStatuses',
    title: 'Contact Lead Status',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'lifecyclestages',
    title: 'Lifecycle Stage',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'owners',
    title: 'HubSpot Owner',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'customObjectMapping',
    title: 'Field Mapping',
    type: 'FIELD_MAPPER_DATA_SOURCE',
    fieldSource: {
      id: 'cachedFields',
      title: 'Field name',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    recordSource: {
      id: 'cacheAllObjectTypes',
      title: 'Custom Object Types',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
  {
    id: 'contactProperties',
    title: 'Contact Properties',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'dealProperties',
    title: 'Deal Properties',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'companyProperties',
    title: 'Company Properties',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'contactLists',
    title: 'Contact Lists',
    type: 'DYNAMIC_DATA_SOURCE',
  },
];

/**
 * triggers for hubspot
 */
exports.triggers = [
  {
    title: 'Record Created',
    name: 'HUBSPOT_TRIGGER_RECORD_CREATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'companies',
          'contacts',
          'deals',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'filterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on new records that match these filters.',
        placeholder: 'field',
        supportedOperators: [
          '$none',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Record Updated',
    name: 'HUBSPOT_TRIGGER_RECORD_UPDATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'companies',
          'contacts',
          'deals',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'filterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on new records that match these filters.',
        placeholder: 'field',
        supportedOperators: [
          '$none',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Record Deleted',
    name: 'HUBSPOT_TRIGGER_RECORD_DELETED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'companies',
          'contacts',
          'deals',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'filterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on new records that match these filters.',
        placeholder: 'field',
        supportedOperators: [
          '$none',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Record Deleted for Privacy (GDPR)',
    name: 'HUBSPOT_TRIGGER_RECORD_DELETED_FOR_PRIVACY',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: ['contact', 'conversation'],
      },
      {
        id: 'filterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on new records that match these filters.',
        placeholder: 'field',
        supportedOperators: [
          '$none',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
];

/**
 * intents for hubspot
 */
exports.intents = [
  {
    title: 'Create Record',
    name: 'HUBSPOT_CREATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'companies',
            dependentInputs: [
              {
                id: 'field-name',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'field-domain',
                title: 'Company Domain Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-hubspot_owner_id',
                title: 'Company Owner',
                type: 'TEXTAREA',
                subtitle:
                  'The HubSpot user who is assigned to this Company. Use Connect Portal Workflow Settings to allow users to select an Owner.',
                placeholder: '{{settings.owner}}',
              },
              {
                id: 'field-lifecyclestage',
                title: 'Lifecycle Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lifecycle Stage if not provided. Use Connect Portal Workflow Settings to allow users to select a Lifecycle Stage.',
                placeholder: '{{settings.lifecycleStage}}',
              },
              {
                id: 'field-hs_lead_status',
                title: 'Lead Status',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lead Status if not provided. Use Connect Portal Workflow Settings to allow users to select a Lead Status.',
                placeholder: '{{settings.leadStatus}}',
              },
              {
                id: 'field-phone',
                title: 'Phone Number',
                type: 'TEXTAREA',
              },
              {
                id: 'field-numberofemployees',
                title: 'Number of Employees',
                type: 'TEXTAREA',
              },
              {
                id: 'field-annualrevenue',
                title: 'Annual Revenue',
                type: 'TEXTAREA',
              },
              {
                id: 'field-description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'additionalFieldsJSON',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'contacts',
            dependentInputs: [
              {
                id: 'field-email',
                title: 'Email',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'field-firstname',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-lastname',
                title: 'Last Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-jobtitle',
                title: 'Job Title',
                type: 'TEXTAREA',
              },
              {
                id: 'field-company',
                title: 'Company Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-hubspot_owner_id',
                title: 'Contact Owner',
                type: 'TEXTAREA',
                subtitle:
                  'The HubSpot user who is assigned to this contact. Use Connect Portal Workflow Settings to allow users to select an Owner.',
                placeholder: '{{settings.owner}}',
              },
              {
                id: 'field-lifecyclestage',
                title: 'Lifecycle Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lifecycle Stage if not provided. Use Connect Portal Workflow Settings to allow users to select a Lifecycle Stage.',
                placeholder: '{{settings.lifecycleStage}}',
              },
              {
                id: 'field-hs_lead_status',
                title: 'Lead Status',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lead Status if not provided. Use Connect Portal Workflow Settings to allow users to select a Lead Status.',
                placeholder: '{{settings.leadStatus}}',
              },
              {
                id: 'additionalFieldsJSON',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'field-dealname',
                title: 'Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'field-amount',
                title: 'Deal Amount',
                type: 'TEXTAREA',
              },
              {
                id: 'field-hubspot_owner_id',
                title: 'Deal Owner',
                type: 'TEXTAREA',
                subtitle:
                  'The HubSpot user who is assigned to this Deal. Use Connect Portal Workflow Settings to allow users to select an Owner.',
                placeholder: '{{settings.owner}}',
              },
              {
                id: 'field-dealstage',
                title: 'Deal Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Deal Stage if not provided. Use Connect Portal Workflow Settings to allow users to select a Deal Stage.',
              },
              {
                id: 'field-description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'field-closedate',
                title: 'Close Date',
                type: 'TEXTAREA',
              },
              {
                id: 'field-hs_next_step',
                title: 'Next Step',
                type: 'TEXTAREA',
              },
              {
                id: 'additionalFieldsJSON',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'engagements',
            dependentInputs: [
              {
                id: 'engagementType',
                title: 'Type',
                type: 'ENUM',
                required: true,
                defaultValue: 'NOTE',
                values: [
                  {
                    value: 'NOTE',
                    dependentInputs: [
                      {
                        id: 'noteBody',
                        title: 'Note',
                        type: 'TEXTAREA',
                        subtitle: 'The body of the note.',
                      },
                    ],
                  },
                  {
                    value: 'EMAIL',
                    dependentInputs: [
                      {
                        id: 'fromEmail',
                        title: 'From email',
                        type: 'TEXTAREA',
                        subtitle: 'Email address of the sender.',
                      },
                      {
                        id: 'fromFirstName',
                        title: 'From first name',
                        type: 'TEXTAREA',
                        subtitle: 'First name of the sender.',
                      },
                      {
                        id: 'fromLastName',
                        title: 'From last name',
                        type: 'TEXTAREA',
                        subtitle: 'Last name of the sender.',
                      },
                      {
                        id: 'toEmail',
                        title: 'To email',
                        type: 'TEXTAREA',
                        subtitle: 'Email address of the recipient.',
                      },
                      {
                        id: 'ccEmail',
                        title: 'CC emails',
                        type: 'TEXTAREA',
                        subtitle:
                          'Array of email addresses for anyone cc’d on the email.',
                      },
                      {
                        id: 'bccEmail',
                        title: 'BCC emails',
                        type: 'TEXTAREA',
                        subtitle:
                          'Array of email addresses for anyone bcc’d on the email.',
                      },
                      {
                        id: 'emailSubject',
                        title: 'Subject',
                        type: 'TEXTAREA',
                        subtitle: 'Subject of the email.',
                      },
                      {
                        id: 'emailText',
                        title: 'Text',
                        type: 'TEXTAREA',
                        subtitle: 'Body of the text-only email.',
                      },
                      {
                        id: 'emailHtml',
                        title: 'HTML',
                        type: 'TEXTAREA',
                        subtitle: 'Body of the HTML email.',
                      },
                    ],
                  },
                  {
                    value: 'CALL',
                    dependentInputs: [
                      {
                        id: 'toNumber',
                        title: 'To number',
                        type: 'TEXTAREA',
                        subtitle: 'The phone number that was called.',
                      },
                      {
                        id: 'fromNumber',
                        title: 'From number',
                        type: 'TEXTAREA',
                        subtitle:
                          'The phone number that was used as the from number.',
                      },
                      {
                        id: 'callNotes',
                        title: 'Body',
                        type: 'TEXTAREA',
                        subtitle: 'Details or notes of the call.',
                      },
                      {
                        id: 'recordingUrl',
                        title: 'Recording URL',
                        type: 'TEXTAREA',
                        subtitle: 'The URL of the call recording.',
                      },
                    ],
                  },
                  {
                    value: 'MEETING',
                    dependentInputs: [
                      {
                        id: 'meetingTitle',
                        title: 'Title',
                        type: 'TEXTAREA',
                        subtitle: 'Title or subject of the meeting.',
                      },
                      {
                        id: 'meetingNotes',
                        title: 'Body',
                        type: 'TEXTAREA',
                        subtitle: 'Details or notes of the meeting.',
                      },
                      {
                        id: 'startTime',
                        title: 'Start time',
                        type: 'TEXTAREA',
                        subtitle:
                          'Start time of the meeting (ISO or Unix timestamp).',
                        placeholder: '2014-04-25T16:15:47-04:00',
                      },
                      {
                        id: 'endTime',
                        title: 'End time',
                        type: 'TEXTAREA',
                        subtitle:
                          'End time of the meeting (ISO or Unix timestamp).',
                        placeholder: '2014-04-25T16:15:47-04:00',
                      },
                    ],
                  },
                  {
                    value: 'TASK',
                    dependentInputs: [
                      {
                        id: 'taskSubject',
                        title: 'Subject',
                        type: 'TEXTAREA',
                        subtitle: 'Title or subject of the task.',
                      },
                      {
                        id: 'taskNotes',
                        title: 'Body',
                        type: 'TEXTAREA',
                        subtitle: 'Details or notes of the task.',
                      },
                      {
                        id: 'taskStatus',
                        title: 'Status',
                        type: 'ENUM',
                        defaultValue: 'NOT_STARTED',
                        values: [
                          'NOT_STARTED',
                          'COMPLETED',
                          'IN_PROGRESS',
                          'WAITING',
                          'DEFERRED',
                        ],
                      },
                    ],
                  },
                ],
              },
              {
                id: 'contactId',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle: 'ID of the Contact the Engagement is related to.',
              },
              {
                id: 'companyId',
                title: 'Company ID',
                type: 'TEXTAREA',
                subtitle: 'ID of the Company the Engagement is related to.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle: 'ID of the owner of the Engagement.',
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'customObjectFields',
                title: 'Record Fields',
                type: 'CODE',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping to specify which fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Update Record',
    name: 'HUBSPOT_UPDATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'companies',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'field-name',
                title: 'Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-domain',
                title: 'Company Domain Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-hubspot_owner_id',
                title: 'Company Owner',
                type: 'TEXTAREA',
                subtitle:
                  'The HubSpot user who is assigned to this Company. Use Connect Portal Workflow Settings to allow users to select an Owner.',
                placeholder: '{{settings.owner}}',
              },
              {
                id: 'field-lifecyclestage',
                title: 'Lifecycle Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lifecycle Stage if not provided. Use Connect Portal Workflow Settings to allow users to select a Lifecycle Stage.',
                placeholder: '{{settings.lifecycleStage}}',
              },
              {
                id: 'field-hs_lead_status',
                title: 'Lead Status',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lead Status if not provided. Use Connect Portal Workflow Settings to allow users to select a Lead Status.',
                placeholder: '{{settings.leadStatus}}',
              },
              {
                id: 'field-phone',
                title: 'Phone Number',
                type: 'TEXTAREA',
              },
              {
                id: 'field-numberofemployees',
                title: 'Number of Employees',
                type: 'TEXTAREA',
              },
              {
                id: 'field-annualrevenue',
                title: 'Annual Revenue',
                type: 'TEXTAREA',
              },
              {
                id: 'field-description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'additionalFieldsJSON',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'contacts',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'field-email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'field-firstname',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-lastname',
                title: 'Last Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-jobtitle',
                title: 'Job Title',
                type: 'TEXTAREA',
              },
              {
                id: 'field-company',
                title: 'Company Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-hubspot_owner_id',
                title: 'Contact Owner',
                type: 'TEXTAREA',
                subtitle:
                  'The HubSpot user who is assigned to this contact. Use Connect Portal Workflow Settings to allow users to select an Owner.',
                placeholder: '{{settings.owner}}',
              },
              {
                id: 'field-lifecyclestage',
                title: 'Lifecycle Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lifecycle Stage if not provided. Use Connect Portal Workflow Settings to allow users to select a Lifecycle Stage.',
                placeholder: '{{settings.lifecycleStage}}',
              },
              {
                id: 'field-hs_lead_status',
                title: 'Lead Status',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Lead Status if not provided. Use Connect Portal Workflow Settings to allow users to select a Lead Status.',
                placeholder: '{{settings.leadStatus}}',
              },
              {
                id: 'additionalFieldsJSON',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'field-dealname',
                title: 'Name',
                type: 'TEXTAREA',
              },
              {
                id: 'field-amount',
                title: 'Deal Amount',
                type: 'TEXTAREA',
              },
              {
                id: 'field-hubspot_owner_id',
                title: 'Deal Owner',
                type: 'TEXTAREA',
                subtitle:
                  'The HubSpot user who is assigned to this Deal. Use Connect Portal Workflow Settings to allow users to select an Owner.',
                placeholder: '{{settings.owner}}',
              },
              {
                id: 'field-dealstage',
                title: 'Deal Stage',
                type: 'TEXTAREA',
                subtitle:
                  'Defaults to the user’s first Deal Stage if not provided. Use Connect Portal Workflow Settings to allow users to select a Deal Stage.',
              },
              {
                id: 'field-description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'field-closedate',
                title: 'Close Date',
                type: 'TEXTAREA',
              },
              {
                id: 'field-hs_next_step',
                title: 'Next Step',
                type: 'TEXTAREA',
              },
              {
                id: 'additionalFieldsJSON',
                title: 'Additional Fields',
                type: 'TEXTAREA',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
              },
            ],
          },
          {
            value: 'engagements',
            dependentInputs: [
              {
                id: 'engagementType',
                title: 'Type',
                type: 'ENUM',
                required: true,
                defaultValue: 'NOTE',
                values: [
                  {
                    value: 'NOTE',
                    dependentInputs: [
                      {
                        id: 'noteBody',
                        title: 'Note',
                        type: 'TEXTAREA',
                        subtitle: 'The body of the note.',
                      },
                    ],
                  },
                  {
                    value: 'EMAIL',
                    dependentInputs: [
                      {
                        id: 'fromEmail',
                        title: 'From email',
                        type: 'TEXTAREA',
                        subtitle: 'Email address of the sender.',
                      },
                      {
                        id: 'fromFirstName',
                        title: 'From first name',
                        type: 'TEXTAREA',
                        subtitle: 'First name of the sender.',
                      },
                      {
                        id: 'fromLastName',
                        title: 'From last name',
                        type: 'TEXTAREA',
                        subtitle: 'Last name of the sender.',
                      },
                      {
                        id: 'toEmail',
                        title: 'To email',
                        type: 'TEXTAREA',
                        subtitle: 'Email address of the recipient.',
                      },
                      {
                        id: 'ccEmail',
                        title: 'CC emails',
                        type: 'TEXTAREA',
                        subtitle:
                          'Array of email addresses for anyone cc’d on the email.',
                      },
                      {
                        id: 'bccEmail',
                        title: 'BCC emails',
                        type: 'TEXTAREA',
                        subtitle:
                          'Array of email addresses for anyone bcc’d on the email.',
                      },
                      {
                        id: 'emailSubject',
                        title: 'Subject',
                        type: 'TEXTAREA',
                        subtitle: 'Subject of the email.',
                      },
                      {
                        id: 'emailText',
                        title: 'Text',
                        type: 'TEXTAREA',
                        subtitle: 'Body of the text-only email.',
                      },
                      {
                        id: 'emailHtml',
                        title: 'HTML',
                        type: 'TEXTAREA',
                        subtitle: 'Body of the HTML email.',
                      },
                    ],
                  },
                  {
                    value: 'CALL',
                    dependentInputs: [
                      {
                        id: 'toNumber',
                        title: 'To number',
                        type: 'TEXTAREA',
                        subtitle: 'The phone number that was called.',
                      },
                      {
                        id: 'fromNumber',
                        title: 'From number',
                        type: 'TEXTAREA',
                        subtitle:
                          'The phone number that was used as the from number.',
                      },
                      {
                        id: 'callNotes',
                        title: 'Body',
                        type: 'TEXTAREA',
                        subtitle: 'Details or notes of the call.',
                      },
                      {
                        id: 'recordingUrl',
                        title: 'Recording URL',
                        type: 'TEXTAREA',
                        subtitle: 'The URL of the call recording.',
                      },
                    ],
                  },
                  {
                    value: 'MEETING',
                    dependentInputs: [
                      {
                        id: 'meetingTitle',
                        title: 'Title',
                        type: 'TEXTAREA',
                        subtitle: 'Title or subject of the meeting.',
                      },
                      {
                        id: 'meetingNotes',
                        title: 'Body',
                        type: 'TEXTAREA',
                        subtitle: 'Details or notes of the meeting.',
                      },
                      {
                        id: 'startTime',
                        title: 'Start time',
                        type: 'TEXTAREA',
                        subtitle:
                          'Start time of the meeting (ISO or Unix timestamp).',
                        placeholder: '2014-04-25T16:15:47-04:00',
                      },
                      {
                        id: 'endTime',
                        title: 'End time',
                        type: 'TEXTAREA',
                        subtitle:
                          'End time of the meeting (ISO or Unix timestamp).',
                        placeholder: '2014-04-25T16:15:47-04:00',
                      },
                    ],
                  },
                  {
                    value: 'TASK',
                    dependentInputs: [
                      {
                        id: 'taskSubject',
                        title: 'Subject',
                        type: 'TEXTAREA',
                        subtitle: 'Title or subject of the task.',
                      },
                      {
                        id: 'taskNotes',
                        title: 'Body',
                        type: 'TEXTAREA',
                        subtitle: 'Details or notes of the task.',
                      },
                      {
                        id: 'taskStatus',
                        title: 'Status',
                        type: 'ENUM',
                        defaultValue: 'NOT_STARTED',
                        values: [
                          'NOT_STARTED',
                          'COMPLETED',
                          'IN_PROGRESS',
                          'WAITING',
                          'DEFERRED',
                        ],
                      },
                    ],
                  },
                ],
              },
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'contactId',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle: 'ID of the Contact the Engagement is related to.',
              },
              {
                id: 'companyId',
                title: 'Company ID',
                type: 'TEXTAREA',
                subtitle: 'ID of the Company the Engagement is related to.',
              },
              {
                id: 'ownerId',
                title: 'Owner ID',
                type: 'TEXTAREA',
                subtitle: 'ID of the owner of the Engagement.',
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to update. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'The ID of the record to update.',
                required: true,
              },
              {
                id: 'customObjectFields',
                title: 'Record Fields',
                type: 'CODE',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping to specify which fields to update.',
                placeholder:
                  '{\n  "property_number": "17",\n  "property_dropdown": "choice_b",\n  "property_radio": "option_1",\n  "property_string": "value",\n  "property_date": "1572480000000"\n}',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get Records',
    name: 'HUBSPOT_GET_RECORDS',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          'companies',
          'contacts',
          'deals',
          {
            value: 'engagements',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Engagements',
                type: 'TEXTAREA',
                subtitle: 'The Name of Engagement you want to fetch.',
                placeholder: 'notes',
                required: true,
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get Record by ID',
    name: 'HUBSPOT_GET_RECORD_BY_ID',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'companies',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'contacts',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'engagements',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Search Records',
    name: 'HUBSPOT_SEARCH_RECORDS',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'companies',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for Records that match specified filters.',
                placeholder: 'field',
                supportedKeys: [
                  'name',
                  'domain',
                  'hubspot_owner_id',
                  'lifecyclestage',
                  'hs_lead_status',
                  'phone',
                  'numberofemployees',
                  'annualrevenue',
                  'description',
                  'createdate',
                  'hs_lastmodifieddate',
                ],
                supportedOperators: [
                  '$none',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$exists',
                  '$doesNotExist',
                ],
              },
            ],
          },
          {
            value: 'contacts',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for Records that match specified filters.',
                placeholder: 'field',
                supportedKeys: [
                  'email',
                  'firstname',
                  'lastname',
                  'jobtitle',
                  'company',
                  'hubspot_owner_id',
                  'lifecyclestage',
                  'hs_lead_status',
                  'createdate',
                  'lastmodifieddate',
                ],
                supportedOperators: [
                  '$none',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$exists',
                  '$doesNotExist',
                ],
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for Records that match specified filters.',
                placeholder: 'field',
                supportedKeys: [
                  'dealname',
                  'amount',
                  'hubspot_owner_id',
                  'dealstage',
                  'description',
                  'closedate',
                  'hs_next_step',
                  'createdate',
                  'hs_lastmodifieddate',
                ],
                supportedOperators: [
                  '$none',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$exists',
                  '$doesNotExist',
                ],
              },
            ],
          },
          {
            value: 'engagements',
            dependentInputs: [
              {
                id: 'engagementType',
                title: 'Engagement Type',
                type: 'ENUM',
                values: [
                  {
                    value: 'CALL',
                    dependentInputs: [
                      {
                        id: 'engagementTypeFilterFormula',
                        title: 'Filter search',
                        type: 'CONDITIONAL',
                        subtitle:
                          'Search for records that match specified filters.',
                        supportedKeys: [
                          'hs_createdate',
                          'hs_lastmodifieddate',
                          'hs_object_id',
                        ],
                        supportedOperators: [
                          '$stringExactlyMatches',
                          '$dateTimeEquals',
                          '$dateTimeBefore',
                          '$dateTimeAfter',
                          '$numberGreaterThanOrEqualTo',
                          '$numberLessThanOrEqualTo',
                          '$numberGreaterThan',
                          '$numberLessThan',
                          '$numberEquals',
                          '$numberDoesNotEqual',
                        ],
                      },
                    ],
                  },
                  {
                    value: 'EMAIL',
                    dependentInputs: [
                      {
                        id: 'engagementTypeFilterFormula',
                        title: 'Filter search',
                        type: 'CONDITIONAL',
                        subtitle:
                          'Search for records that match specified filters.',
                        supportedKeys: [
                          'hs_createdate',
                          'hs_lastmodifieddate',
                          'hs_object_id',
                        ],
                        supportedOperators: [
                          '$stringExactlyMatches',
                          '$dateTimeEquals',
                          '$dateTimeBefore',
                          '$dateTimeAfter',
                          '$numberGreaterThanOrEqualTo',
                          '$numberLessThanOrEqualTo',
                          '$numberGreaterThan',
                          '$numberLessThan',
                          '$numberEquals',
                          '$numberDoesNotEqual',
                        ],
                      },
                    ],
                  },
                  {
                    value: 'MEETING',
                    dependentInputs: [
                      {
                        id: 'engagementTypeFilterFormula',
                        title: 'Filter search',
                        type: 'CONDITIONAL',
                        subtitle:
                          'Search for records that match specified filters.',
                        supportedKeys: [
                          'hs_createdate',
                          'hs_lastmodifieddate',
                          'hs_object_id',
                        ],
                        supportedOperators: [
                          '$stringExactlyMatches',
                          '$dateTimeEquals',
                          '$dateTimeBefore',
                          '$dateTimeAfter',
                          '$numberGreaterThanOrEqualTo',
                          '$numberLessThanOrEqualTo',
                          '$numberGreaterThan',
                          '$numberLessThan',
                          '$numberEquals',
                          '$numberDoesNotEqual',
                        ],
                      },
                    ],
                  },
                  {
                    value: 'NOTE',
                    dependentInputs: [
                      {
                        id: 'engagementTypeFilterFormula',
                        title: 'Filter search',
                        type: 'CONDITIONAL',
                        subtitle:
                          'Search for records that match specified filters.',
                        supportedKeys: [
                          'hs_createdate',
                          'hs_lastmodifieddate',
                          'hs_object_id',
                        ],
                        supportedOperators: [
                          '$stringExactlyMatches',
                          '$dateTimeEquals',
                          '$dateTimeBefore',
                          '$dateTimeAfter',
                          '$numberGreaterThanOrEqualTo',
                          '$numberLessThanOrEqualTo',
                          '$numberGreaterThan',
                          '$numberLessThan',
                          '$numberEquals',
                          '$numberDoesNotEqual',
                        ],
                      },
                    ],
                  },
                  {
                    value: 'TASK',
                    dependentInputs: [
                      {
                        id: 'engagementTypeFilterFormula',
                        title: 'Filter search',
                        type: 'CONDITIONAL',
                        subtitle:
                          'Search for records that match specified filters.',
                        supportedKeys: [
                          'hs_createdate',
                          'hs_lastmodifieddate',
                          'hs_object_id',
                        ],
                        supportedOperators: [
                          '$stringExactlyMatches',
                          '$dateTimeEquals',
                          '$dateTimeBefore',
                          '$dateTimeAfter',
                          '$numberGreaterThanOrEqualTo',
                          '$numberLessThanOrEqualTo',
                          '$numberGreaterThan',
                          '$numberLessThan',
                          '$numberEquals',
                          '$numberDoesNotEqual',
                        ],
                      },
                    ],
                  },
                ],
              },
              {
                id: 'engagementFilterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for records that match specified filters.',
                supportedKeys: [
                  'contactIds',
                  'companyIds',
                  'dealIds',
                  'ownerIds',
                  'ticketIds',
                ],
                supportedOperators: ['$stringExactlyMatches'],
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'filterFormula',
                title: 'Filter search',
                type: 'CONDITIONAL',
                subtitle: 'Search for Records that match specified filters.',
                placeholder: 'field',
                supportedOperators: [
                  '$none',
                  '$stringExactlyMatches',
                  '$stringDoesNotExactlyMatch',
                  '$numberGreaterThan',
                  '$numberLessThan',
                  '$numberEquals',
                  '$numberDoesNotEqual',
                  '$dateTimeAfter',
                  '$dateTimeBefore',
                  '$dateTimeEquals',
                  '$exists',
                  '$doesNotExist',
                ],
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Delete Record by ID',
    name: 'HUBSPOT_DELETE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record Type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'companies',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'contacts',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'deals',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: 'engagements',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Custom Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Name of the Custom Object to create. Use Connect Portal Workflow Settings to allow users to select a Custom Object Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                required: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get Contacts by List ID',
    name: 'HUBSPOT_GET_CONTACTS_BY_LIST_ID',
    output: [],
    inputs: [
      {
        id: 'listId',
        title: 'List ID',
        type: 'TEXTAREA',
        subtitle: 'The ID of the list to get contacts from.',
        placeholder: '{{settings.listId}}',
        required: true,
      },
    ],
  },
];

/**
 * integration class for hubspot
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'hubspot',

  /**
   * action step methods
   */
  actions: {
    /**
     * Create Record
     * method for HUBSPOT_CREATE_RECORD
     */
    createRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'HUBSPOT_CREATE_RECORD',
        actionType: 'hubspot',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update Record
     * method for HUBSPOT_UPDATE_RECORD
     */
    updateRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'HUBSPOT_UPDATE_RECORD',
        actionType: 'hubspot',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Records
     * method for HUBSPOT_GET_RECORDS
     */
    getRecords: (inputs, options) => {
      return new ActionStep({
        intent: 'HUBSPOT_GET_RECORDS',
        actionType: 'hubspot',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Record by ID
     * method for HUBSPOT_GET_RECORD_BY_ID
     */
    getRecordById: (inputs, options) => {
      return new ActionStep({
        intent: 'HUBSPOT_GET_RECORD_BY_ID',
        actionType: 'hubspot',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Search Records
     * method for HUBSPOT_SEARCH_RECORDS
     */
    searchRecords: (inputs, options) => {
      return new ActionStep({
        intent: 'HUBSPOT_SEARCH_RECORDS',
        actionType: 'hubspot',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Delete Record by ID
     * method for HUBSPOT_DELETE_RECORD
     */
    deleteRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'HUBSPOT_DELETE_RECORD',
        actionType: 'hubspot',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get Contacts by List ID
     * method for HUBSPOT_GET_CONTACTS_BY_LIST_ID
     */
    getContactsByListId: (inputs, options) => {
      return new ActionStep({
        intent: 'HUBSPOT_GET_CONTACTS_BY_LIST_ID',
        actionType: 'hubspot',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {
    /**
     * Record Created
     * method for HUBSPOT_TRIGGER_RECORD_CREATED
     */
    recordCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'HUBSPOT_TRIGGER_RECORD_CREATED',
        actionType: 'hubspot',
        parameters: inputs,
      });
    },
    /**
     * Record Updated
     * method for HUBSPOT_TRIGGER_RECORD_UPDATED
     */
    recordUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'HUBSPOT_TRIGGER_RECORD_UPDATED',
        actionType: 'hubspot',
        parameters: inputs,
      });
    },
    /**
     * Record Deleted
     * method for HUBSPOT_TRIGGER_RECORD_DELETED
     */
    recordDeleted: (inputs) => {
      return new ActionTriggerStep({
        intent: 'HUBSPOT_TRIGGER_RECORD_DELETED',
        actionType: 'hubspot',
        parameters: inputs,
      });
    },
    /**
     * Record Deleted for Privacy (GDPR)
     * method for HUBSPOT_TRIGGER_RECORD_DELETED_FOR_PRIVACY
     */
    recordDeletedForPrivacy: (inputs) => {
      return new ActionTriggerStep({
        intent: 'HUBSPOT_TRIGGER_RECORD_DELETED_FOR_PRIVACY',
        actionType: 'hubspot',
        parameters: inputs,
      });
    },
  },
};
