// @useparagon/core/dist/triggers/endpoint/endpoint.step.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndpointStep = void 0;
const execution_1 = require("../../execution");
const resolvers_1 = require("../../resolvers");
const step_interface_1 = require("../../steps/step.interface");
const trigger_1 = require("../trigger");
class EndpointStep extends trigger_1.TriggerStep {
    type = step_interface_1.StepType.ENDPOINT;
    _parameters;
    parameters;
    get output() {
        return (0, execution_1.createProxyInputReference)(this, () => this.id, execution_1.ProxyVariableSourceType.STEP_REFERENCE);
    }
    constructor(params) {
        super(params);
        this._parameters = params;
        this.parameters = this.serializeParameters();
    }
    serializeParameters() {
        const initParams = this._parameters;
        const defaultParams = {
            httpMethod: 'POST',
            paramValidations: [],
            headerValidations: [],
            bodyValidations: [],
            allowArbitraryPayload: false,
            path: this.dependencies?.workflowId,
        };
        if (initParams.allowArbitraryPayload) {
            return {
                ...defaultParams,
                objectMapping: this._parameters.objectMapping
                    ? (0, resolvers_1.resolveToTokenizedSource)(this._parameters.objectMapping, resolvers_1.DataType.STRING, this.dependencies)
                    : undefined,
            };
        }
        return { ...defaultParams, ...initParams };
    }
}
exports.EndpointStep = EndpointStep;
