// @useparagon/core/dist/execution/context.interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxyVariableSourceType = void 0;
var ProxyVariableSourceType;
(function (ProxyVariableSourceType) {
    ProxyVariableSourceType["STEP_REFERENCE"] = "STEP_REFERENCE";
    ProxyVariableSourceType["SECRET_REFERENCE"] = "SECRET_REFERENCE";
    ProxyVariableSourceType["PERSONA_META_REFERENCE"] = "PERSONA_META_REFERENCE";
    ProxyVariableSourceType["PERSONA_ENDUSERID_REFERENCE"] = "PERSONA_ENDUSERID_REFERENCE";
    ProxyVariableSourceType["WORKFLOW_INPUT_REFERENCE"] = "WORKFLOW_INPUT_REFERENCE";
    ProxyVariableSourceType["CONNECT_USER_PROVIDER_ID"] = "CONNECT_USER_PROVIDER_ID";
    ProxyVariableSourceType["CONNECT_USER_PROVIDER_DATA"] = "CONNECT_USER_PROVIDER_DATA";
    ProxyVariableSourceType["SAME_STEP_REFERENCE"] = "SAME_STEP_REFERENCE";
    ProxyVariableSourceType["OAUTH_ACCESS_TOKEN"] = "OAUTH_ACCESS_TOKEN";
    ProxyVariableSourceType["CONNECT_USER_CREDENTIAL_ID"] = "CONNECT_USER_CREDENTIAL_ID";
    ProxyVariableSourceType["CONNECT_USER_CONFIGURATION_ID"] = "CONNECT_USER_CONFIGURATION_ID";
    ProxyVariableSourceType["CUSTOM_WEBHOOK_TRIGGER_URL"] = "CUSTOM_WEBHOOK_TRIGGER_URL";
    ProxyVariableSourceType["CUSTOM_WEBHOOK_TRIGGER_SETUP_RESPONSE"] = "CUSTOM_WEBHOOK_TRIGGER_SETUP_RESPONSE";
    ProxyVariableSourceType["CUSTOM_WEBHOOK_TRIGGER_HANDSHAKE_RESPONSE"] = "CUSTOM_WEBHOOK_TRIGGER_HANDSHAKE_RESPONSE";
    ProxyVariableSourceType["CUSTOM_WEBHOOK_TRIGGER_WEBHOOK_RESPONSE"] = "CUSTOM_WEBHOOK_TRIGGER_WEBHOOK_RESPONSE";
    ProxyVariableSourceType["CREDENTIAL_VALUES"] = "CREDENTIAL_VALUES";
    ProxyVariableSourceType["PROFILE_CONFIG_KEY"] = "PROFILE_CONFIG_KEY";
})(ProxyVariableSourceType || (exports.ProxyVariableSourceType = ProxyVariableSourceType = {}));
