// @useparagon/core/dist/integration/integration.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSlugForIntegration = exports.convertSerializedInputsToUserDefinedConfigInputs = exports.isCustomIntegration = void 0;
const inputs_1 = require("../inputs");
const isCustomIntegration = (integration) => integration.startsWith('custom.') || integration === 'undefined';
exports.isCustomIntegration = isCustomIntegration;
const convertSerializedInputsToUserDefinedConfigInputs = (inputs) => {
    return inputs.map((input) => ({
        id: input.id,
        type: input.type === inputs_1.SidebarInputType.Password ? 'password' : 'text',
        title: input.title,
        subtitle: input.subtitle || '',
        placeholder: input.placeholder || '',
        suffixLabel: input.suffixLabel || '',
    }));
};
exports.convertSerializedInputsToUserDefinedConfigInputs = convertSerializedInputsToUserDefinedConfigInputs;
function generateSlugForIntegration(customIntegration) {
    const slug = customIntegration.name
        ? customIntegration.name
            .replace(/[\W_]+/g, '')
            .trim()
            .toLowerCase()
        : undefined;
    if (!slug) {
        throw new Error(`Could not generate slug for custom integration with name : ${customIntegration.name}`);
    }
    return `custom.${slug}`;
}
exports.generateSlugForIntegration = generateSlugForIntegration;
