// @useparagon/integrations/zohocrm/index.js
const { ActionStep, ActionTriggerStep } = require('@useparagon/core');

/**
 * this should be used whenever we are creating inputs for integrations/workflows
 */
exports.createInputs = (inputs) => inputs;

/**
 * inputs for zohocrm
 */
exports.inputs = [
  {
    id: 'recordTypeCacheKey',
    title: 'Record Type',
    type: 'DYNAMIC_DATA_SOURCE',
  },
  {
    id: 'customItemMapping',
    title: 'Field Mapping',
    type: 'FIELD_MAPPER_DATA_SOURCE',
    fieldSource: {
      id: 'fieldsCacheKey',
      title: 'Field',
      type: 'DYNAMIC_DATA_SOURCE',
    },
    recordSource: {
      id: 'recordTypeCacheKey',
      title: 'Record Type',
      type: 'DYNAMIC_DATA_SOURCE',
    },
  },
];

/**
 * triggers for zohocrm
 */
exports.triggers = [
  {
    title: 'New Record',
    name: 'ZOHO_CRM_TRIGGER_RECORD_CREATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'Accounts',
          'Contacts',
          'Leads',
          'Tasks',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to create. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'recordsFilterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on new records that match these filters.',
        supportedKeys: ['id'],
        supportedOperators: [
          '$none',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringDoesNotEndWith',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$booleanTrue',
          '$booleanFalse',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Record Updated',
    name: 'ZOHO_CRM_TRIGGER_RECORD_UPDATED',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'Accounts',
          'Contacts',
          'Leads',
          'Tasks',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to update. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'recordsFilterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Only trigger on updated records that match these filters.',
        supportedKeys: ['id'],
        supportedOperators: [
          '$none',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringDoesNotEndWith',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$booleanTrue',
          '$booleanFalse',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
];

/**
 * intents for zohocrm
 */
exports.intents = [
  {
    title: 'Create record',
    name: 'ZOHO_CRM_CREATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Accounts',
            dependentInputs: [
              {
                id: 'Account_Name',
                title: 'Accout Name',
                type: 'TEXTAREA',
                subtitle: 'Specify the Account Name.',
                required: true,
              },
              {
                id: 'Website',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'Phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'Owner',
                title: 'Owner ID',
                type: 'TEXTAREA',
              },
              {
                id: 'accountAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Account fields to update.',
                placeholder:
                  '{\n  "Annual_Revenue": 850000,\n  "Shipping_Street": null,\n  "Ownership": "Partnership",\n}',
              },
            ],
          },
          {
            value: 'Contacts',
            dependentInputs: [
              {
                id: 'First_Name',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'Last_Name',
                title: 'Last Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'Account_Name',
                title: 'Accout Name',
                type: 'TEXTAREA',
                subtitle: 'Specify the Account Name.',
              },
              {
                id: 'Email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'Title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle:
                  'Title of the contact, such as CEO or Vice President.',
              },
              {
                id: 'Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'contactAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "Twitter": "krismarrier_sample",\n  "Other_Zip": null,\n  "Mailing_Street": "228 Runamuck Pl #2808",\n}',
              },
            ],
          },
          {
            value: 'Leads',
            dependentInputs: [
              {
                id: 'First_Name',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'Last_Name',
                title: 'Last Name',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'Company',
                title: 'Company',
                type: 'TEXTAREA',
              },
              {
                id: 'Email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'Phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'Website',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'Designation',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle:
                  'Title of the contact, such as CEO or Vice President.',
              },
              {
                id: 'Lead_Status',
                title: 'Lead Status',
                type: 'TEXTAREA',
              },
              {
                id: 'Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'leadAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Lead fields to update.',
                placeholder: '{\n  "State": C.A,\n  "Skype_ID": skype_12id,\n}',
              },
            ],
          },
          {
            value: 'Tasks',
            dependentInputs: [
              {
                id: 'Who_Id',
                title: 'Contact ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Contact this Task is related to. Specify this when you want to associate a Task with a Contact.',
              },
              {
                id: 'What_Id',
                title: 'Related to Record ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the related record the Task is associated with. Specify this when you want to associate a Task with an Account, Lead or other module record.',
              },
              {
                id: '$se_module',
                title: 'Associated Module Name',
                type: 'TEXTAREA',
                subtitle:
                  'The Zoho CRM API name of the parent module the Task is associated with. This key is mandatory when including a Contact ID or Related to Record ID.',
                placeholder: 'Accounts',
              },
              {
                id: 'Subject',
                title: 'Subject',
                type: 'TEXTAREA',
                required: true,
              },
              {
                id: 'Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'Status',
                title: 'Status',
                type: 'ENUM',
                values: ['Not Started', 'In Progress', 'Completed'],
              },
              {
                id: 'taskAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Task fields to update.',
                placeholder:
                  '{\n  "Due_Date": "2022-03-18",\n  "Priority": "Highest",\n}',
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to create. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'additionalFields',
                title: 'Record Fields',
                type: 'CODE',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Field Mapping to specify which fields to update.',
                placeholder:
                  '{\n  "Annual_Revenue": 850000,\n  "Shipping_Street": null,\n  "Ownership": "Partnership",\n}',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Update record',
    name: 'ZOHO_CRM_UPDATE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          {
            value: 'Accounts',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the record ID to update.',
                required: true,
              },
              {
                id: 'Account_Name',
                title: 'Accout Name',
                type: 'TEXTAREA',
                subtitle: 'Specify the Account Name.',
              },
              {
                id: 'Website',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'Phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'Owner',
                title: 'Owner ID',
                type: 'TEXTAREA',
              },
              {
                id: 'accountAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Account fields to update.',
                placeholder:
                  '{\n  "Annual_Revenue": 850000,\n  "Shipping_Street": null,\n  "Ownership": "Partnership",\n}',
              },
            ],
          },
          {
            value: 'Contacts',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the record ID to update.',
                required: true,
              },
              {
                id: 'First_Name',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'Last_Name',
                title: 'Last Name',
                type: 'TEXTAREA',
              },
              {
                id: 'Account_Name',
                title: 'Accout Name',
                type: 'TEXTAREA',
                subtitle: 'Specify the Account Name.',
              },
              {
                id: 'Email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'Title',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle:
                  'Title of the contact, such as CEO or Vice President.',
              },
              {
                id: 'Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'contactAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Contact fields to update.',
                placeholder:
                  '{\n  "Twitter": "krismarrier_sample",\n  "Other_Zip": null,\n  "Mailing_Street": "228 Runamuck Pl #2808",\n}',
              },
            ],
          },
          {
            value: 'Leads',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the record ID to update.',
                required: true,
              },
              {
                id: 'First_Name',
                title: 'First Name',
                type: 'TEXTAREA',
              },
              {
                id: 'Last_Name',
                title: 'Last Name',
                type: 'TEXTAREA',
              },
              {
                id: 'Company',
                title: 'Company',
                type: 'TEXTAREA',
              },
              {
                id: 'Email',
                title: 'Email',
                type: 'TEXTAREA',
              },
              {
                id: 'Phone',
                title: 'Phone',
                type: 'TEXTAREA',
              },
              {
                id: 'Website',
                title: 'Website',
                type: 'TEXTAREA',
              },
              {
                id: 'Designation',
                title: 'Title',
                type: 'TEXTAREA',
                subtitle:
                  'Title of the contact, such as CEO or Vice President.',
              },
              {
                id: 'Lead_Status',
                title: 'Lead Status',
                type: 'TEXTAREA',
              },
              {
                id: 'Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'leadAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Lead fields to update.',
                placeholder: '{\n  "State": C.A,\n  "Skype_ID": skype_12id,\n}',
              },
            ],
          },
          {
            value: 'Tasks',
            dependentInputs: [
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the record ID to update.',
                required: true,
              },
              {
                id: 'Who_Id',
                title: 'Name ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the Contact or Lead this Task is related to.',
              },
              {
                id: 'What_Id',
                title: 'Related to ID',
                type: 'TEXTAREA',
                subtitle:
                  'The ID of the account the contact is associated with. Specify this only when the Name ID is a contact.',
              },
              {
                id: '$se_module',
                title: 'Associated Module Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API name of the parent module that the contact is associated with. This key is mandatory when you include Who_id and What_Id.',
                placeholder: 'Accounts',
              },
              {
                id: 'Subject',
                title: 'Subject',
                type: 'TEXTAREA',
              },
              {
                id: 'Description',
                title: 'Description',
                type: 'TEXTAREA',
              },
              {
                id: 'Status',
                title: 'Status',
                type: 'ENUM',
                values: ['Not Started', 'In Progress', 'Completed'],
              },
              {
                id: 'taskAdditionalFields',
                title: 'Additional Fields',
                type: 'CODE',
                subtitle:
                  'Specify any other fields that should be updated in JSON below. Use Connect Portal Workflow Settings to allow users to select which Task fields to update.',
                placeholder:
                  '{\n  "Due_Date": "2022-03-18",\n  "Priority": "Highest",\n}',
              },
            ],
          },
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to update. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
              {
                id: 'recordId',
                title: 'Record ID',
                type: 'TEXTAREA',
                subtitle: 'Specify the record ID to update.',
                required: true,
              },
              {
                id: 'additionalFields',
                title: 'Record Fields',
                type: 'CODE',
                subtitle:
                  'Use Connect Portal Workflow Settings to allow users to select a Field Mapping to specify which fields to update.',
                placeholder:
                  '{\n  "Annual_Revenue": 850000,\n  "Shipping_Street": null,\n  "Ownership": "Partnership",\n}',
              },
            ],
          },
        ],
      },
    ],
  },
  {
    title: 'Get record by ID',
    name: 'ZOHO_CRM_GET_RECORD_BY_ID',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'Accounts',
          'Contacts',
          'Leads',
          'Tasks',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to get record. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'recordId',
        title: 'Record ID',
        type: 'TEXTAREA',
        subtitle: 'Specify the record ID to fetch.',
        required: true,
      },
    ],
  },
  {
    title: 'Search Records',
    name: 'ZOHO_CRM_SEARCH_RECORDS',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'Accounts',
          'Contacts',
          'Leads',
          'Tasks',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to search. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'filterFormula',
        title: 'Filter search',
        type: 'CONDITIONAL',
        subtitle: 'Search for records that match specified filters.',
        supportedOperators: [
          '$none',
          '$stringContains',
          '$stringDoesNotContain',
          '$stringExactlyMatches',
          '$stringDoesNotExactlyMatch',
          '$stringIsIn',
          '$stringIsNotIn',
          '$stringStartsWith',
          '$stringDoesNotStartWith',
          '$stringEndsWith',
          '$stringDoesNotEndWith',
          '$numberGreaterThan',
          '$numberLessThan',
          '$numberEquals',
          '$numberDoesNotEqual',
          '$dateTimeAfter',
          '$dateTimeBefore',
          '$dateTimeEquals',
          '$booleanTrue',
          '$booleanFalse',
          '$exists',
          '$doesNotExist',
        ],
      },
    ],
  },
  {
    title: 'Search records by COQL query',
    name: 'ZOHO_CRM_WRITE_COQL_QUERY',
    output: [],
    inputs: [
      {
        id: 'query',
        title: 'COQL Query',
        type: 'CODE',
        placeholder:
          "SELECT Account_Name, Description\nFROM Accounts\nWHERE Name = 'Ethan'",
        required: true,
      },
    ],
  },
  {
    title: 'Delete Record',
    name: 'ZOHO_CRM_DELETE_RECORD',
    output: [],
    inputs: [
      {
        id: 'recordType',
        title: 'Record type',
        type: 'ENUM',
        required: true,
        values: [
          'Accounts',
          'Contacts',
          'Leads',
          'Tasks',
          {
            value: '<specified by objectName property>',
            dependentInputs: [
              {
                id: 'objectName',
                title: 'Object Name',
                type: 'TEXTAREA',
                subtitle:
                  'The API Name of the Object to delete. Use Connect Portal Workflow Settings to allow users to select a Field Mapping.',
                placeholder: '{{settings.objectType}}',
                required: true,
              },
            ],
          },
        ],
      },
      {
        id: 'recordId',
        title: 'Record ID',
        type: 'TEXTAREA',
        subtitle: 'Specify the record ID to delete.',
        required: true,
      },
    ],
  },
];

/**
 * integration class for zohocrm
 */
exports.integration = {
  /**
   * name of integration
   */
  name: 'zohocrm',

  /**
   * action step methods
   */
  actions: {
    /**
     * Create record
     * method for ZOHO_CRM_CREATE_RECORD
     */
    zohoCrmCreateRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'ZOHO_CRM_CREATE_RECORD',
        actionType: 'zohocrm',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Update record
     * method for ZOHO_CRM_UPDATE_RECORD
     */
    zohoCrmUpdateRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'ZOHO_CRM_UPDATE_RECORD',
        actionType: 'zohocrm',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Get record by ID
     * method for ZOHO_CRM_GET_RECORD_BY_ID
     */
    zohoCrmGetRecordById: (inputs, options) => {
      return new ActionStep({
        intent: 'ZOHO_CRM_GET_RECORD_BY_ID',
        actionType: 'zohocrm',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Search Records
     * method for ZOHO_CRM_SEARCH_RECORDS
     */
    zohoCrmSearchRecords: (inputs, options) => {
      return new ActionStep({
        intent: 'ZOHO_CRM_SEARCH_RECORDS',
        actionType: 'zohocrm',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Search records by COQL query
     * method for ZOHO_CRM_WRITE_COQL_QUERY
     */
    zohoCrmWriteCoqlQuery: (inputs, options) => {
      return new ActionStep({
        intent: 'ZOHO_CRM_WRITE_COQL_QUERY',
        actionType: 'zohocrm',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
    /**
     * Delete Record
     * method for ZOHO_CRM_DELETE_RECORD
     */
    zohoCrmDeleteRecord: (inputs, options) => {
      return new ActionStep({
        intent: 'ZOHO_CRM_DELETE_RECORD',
        actionType: 'zohocrm',
        parameters: inputs,
        description: options.description,
        autoRetry: options.autoRetry,
        continueWorkflowOnError: options.continueWorkflowOnError,
      });
    },
  },

  /**
   * trigger step methods
   */
  triggers: {
    /**
     * New Record
     * method for ZOHO_CRM_TRIGGER_RECORD_CREATED
     */
    zohoCrmTriggerRecordCreated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'ZOHO_CRM_TRIGGER_RECORD_CREATED',
        actionType: 'zohocrm',
        parameters: inputs,
      });
    },
    /**
     * Record Updated
     * method for ZOHO_CRM_TRIGGER_RECORD_UPDATED
     */
    zohoCrmTriggerRecordUpdated: (inputs) => {
      return new ActionTriggerStep({
        intent: 'ZOHO_CRM_TRIGGER_RECORD_UPDATED',
        actionType: 'zohocrm',
        parameters: inputs,
      });
    },
  },
};
